diff --git a/data/io.elementary.settings-daemon.service b/data/io.elementary.settings-daemon.service
new file mode 100644
index 0000000..93ae29b
--- /dev/null
+++ b/data/io.elementary.settings-daemon.service
@@ -0,0 +1,12 @@
+[Unit]
+Description=Settings daemon for pantheon desktop
+
+[Service]
+Type=simple
+StandardError=journal
+ExecStart=/usr/bin/io.elementary.settings-daemon
+Restart=on-failure
+
+[Install]
+WantedBy=gnome-session@pantheon-wayland.target
+WantedBy=gnome-session@pantheon.target
\ No newline at end of file
diff --git a/data/meson.build b/data/meson.build
index 22c3103..d2c76f6 100644
--- a/data/meson.build
+++ b/data/meson.build
@@ -87,4 +87,9 @@ if systemd_system_unit_dir != 'no' and systemd_user_unit_dir != 'no'
         meson.project_name() + '.system-update.timer',
         install_dir: systemd_user_unit_dir
     )
+
+    install_data(
+        meson.project_name() + '.service',
+        install_dir: systemd_user_unit_dir
+    )
 endif
