diff --git a/vapi/Clutter-51.metadata b/vapi/Clutter-51.metadata
new file mode 100644
index 0000000..f7f4205
--- /dev/null
+++ b/vapi/Clutter-51.metadata
@@ -0,0 +1,145 @@
+// Non mini-object
+ActorBox struct
+Margin struct
+PaintVolume struct
+Perspective struct
+
+*.ref unowned
+* cheader_filename="clutter/clutter.h"
+
+// Fix the few clutter-pango headers
+Text cheader_filename="clutter/clutter-pango.h"
+TextBuffer cheader_filename="clutter/clutter-pango.h"
+TextNode cheader_filename="clutter/clutter-pango.h"
+Actor
+	.get_pango_context cheader_filename="clutter/clutter-pango.h"
+	.create_pango_context cheader_filename="clutter/clutter-pango.h"
+	.create_pango_layout cheader_filename="clutter/clutter-pango.h"
+
+
+Actor
+	.apply_transform.matrix ref
+	.get_abs_allocation_vertices.verts out=false
+Event.type#method name="get_type"
+
+// ???
+Actor.has_pointer#method name="get_has_pointer"
+
+// Not all backing symbols are deprecated
+Actor.pick deprecated=false
+
+// Nullable return values
+Actor
+	.get_parent nullable
+
+// The original CType has been overridden by the annotations
+Actor
+	.allocate.box ctype="const ClutterActorBox *"
+	.get_stage ctype="ClutterActor *"
+Action.handle_event.event ctype="const ClutterEvent *"
+
+// method/virtual-method/signal don't match
+Actor
+	.event#method name="emit_event"
+	.get_paint_volume#virtual_method name="get_paint_volume_vfunc"
+	.get_paint_volume#virtual_method.volume out
+Text
+	.activate#method name="try_activate"
+	.insert_text#signal skip
+TextBuffer.get_text#virtual_method name="get_text_with_length"
+
+// Default values
+Stage.read_pixels
+	.width default=-1
+	.height default=-1
+Stage.paint_to_buffer
+    .data type="uint8[]"
+Text
+	.position_to_coords.line_height default=null
+
+// Skipped by g-i for unknown reasons
+LayoutManager
+	.create_child_meta skip=false
+
+// We can use static strings
+PaintNode
+	.set_static_name skip=false
+
+// Variadic arguments
+Backend
+	.get_cogl_context skip=false
+Interval
+	.new skip=false
+	.get_interval skip=false
+	.set_final skip=false
+	.set_initial skip=false
+	.set_interval skip=false
+LayoutManager
+	.child_get skip=false
+	.child_set skip=false
+
+// Skipped upstream for unknown reasons
+Interval.register_progress_func skip=false
+
+// struct/class confusion
+ActorBox
+	.new skip
+	.from_vertices skip
+Margin
+	.new skip
+
+// Upstream
+Event
+	.get_position.position out
+
+FrameListenerIface skip
+FrameClock.new skip
+
+// Remove for clutter-2.0
+/////////////////////////
+
+StageView.layout skip
+
+Stage
+	.paint_view.redraw_clip type="Cairo.Region"
+
+// *Event should be compact classes derived from Clutter.Event
+Event.type skip=false
+AnyEvent struct=false base_type="Clutter.Event"
+ButtonEvent struct=false base_type="Clutter.Event"
+CrossingEvent struct=false base_type="Clutter.Event"
+DeviceEvent struct=false base_type="Clutter.Event"
+IMEvent struct=false base_type="Clutter.Event"
+KeyEvent struct=false base_type="Clutter.Event"
+MotionEvent struct=false base_type="Clutter.Event"
+PadButtonEvent struct=false base_type="Clutter.Event"
+PadRingEvent struct=false base_type="Clutter.Event"
+PadStripEvent struct=false base_type="Clutter.Event"
+ProximityEvent struct=false base_type="Clutter.Event"
+ScrollEvent struct=false base_type="Clutter.Event"
+TouchEvent struct=false base_type="Clutter.Event"
+TouchpadHoldEvent struct=false base_type="Clutter.Event"
+TouchpadPinchEvent struct=false base_type="Clutter.Event"
+TouchpadSwipeEvent struct=false base_type="Clutter.Event"
+
+// Keysyms used to be CLUTTER_X instead of CLUTTER_KEY_X
+*#constant skip
+CURRENT_TIME skip=false
+PRIORITY_REDRAW skip=false
+
+// Clutter devs don't like us creating nested namespaces
+value_* name="value_(.+)" parent="Clutter.Value"
+threads_* name="threads_(.+)" parent="Clutter.Threads"
+
+// There is no way to know sealed classes before GLib 2.70
+ColorState sealed
+FrameClock sealed
+TextureContent sealed
+
+TextureContent.new_from_texture symbol_type="constructor"
+
+// Possibly keep
+KEY_* skip=false name="KEY_(.+)" type="uint" parent="Clutter.Key"
+BUTTON_* skip=false name="BUTTON_(.+)" type="uint32" parent="Clutter.Button"
+EVENT_STOP skip=false type="bool"
+EVENT_PROPAGATE skip=false type="bool"
diff --git a/vapi/Cogl-51-custom.vala b/vapi/Cogl-51-custom.vala
new file mode 100644
index 0000000..90d46b3
--- /dev/null
+++ b/vapi/Cogl-51-custom.vala
@@ -0,0 +1,66 @@
+namespace Cogl {
+	public struct Color {
+		[CCode (cname="cogl_color_init_from_4f")]
+		public Color.from_4f (float red, float green, float blue, float alpha);
+		[CCode (cname="cogl_color_init_from_hsl")]
+		public Color.from_hsl (float hue, float saturation, float luminance);
+		[CCode (cname = "_vala_cogl_color_from_string")]
+		public static Cogl.Color? from_string (string str) {
+			Cogl.Color color = {};
+			if (color.init_from_string (str))
+				return color;
+
+			return null;
+		}
+	}
+	[CCode (cheader_filename = "cogl/cogl.h", has_type_id = false)]
+	public struct VertexP2 {
+		public float x;
+		public float y;
+	}
+	[CCode (cheader_filename = "cogl/cogl.h", has_type_id = false)]
+	public struct VertexP2C4 {
+		public float x;
+		public float y;
+		public uint8 r;
+		public uint8 g;
+		public uint8 b;
+		public uint8 a;
+	}
+	[CCode (cheader_filename = "cogl/cogl.h", has_type_id = false)]
+	public struct VertexP2T2 {
+		public float x;
+		public float y;
+		public float s;
+		public float t;
+	}
+	[CCode (cheader_filename = "cogl/cogl.h", has_type_id = false)]
+	public struct VertexP2T2C4 {
+		public float x;
+		public float y;
+		public float s;
+		public float t;
+		public uint8 r;
+		public uint8 g;
+		public uint8 b;
+		public uint8 a;
+	}
+	[CCode (cheader_filename = "cogl/cogl.h", has_type_id = false)]
+	public struct VertexP3 {
+		public float x;
+		public float y;
+		public float z;
+	}
+	[CCode (cheader_filename = "cogl/cogl.h", has_type_id = false)]
+	public struct VertexP3T2 {
+		public float x;
+		public float y;
+		public float z;
+		public float s;
+		public float t;
+	}
+	[CCode (cheader_filename = "cogl/cogl.h", cprefix = "COGL_PIXEL_FORMAT_", type_id = "cogl_pixel_format_get_type ()")]
+	public enum PixelFormat {
+		CAIRO_ARGB32_COMPAT;
+	}
+}
diff --git a/vapi/Cogl-51.metadata b/vapi/Cogl-51.metadata
new file mode 100644
index 0000000..f66d6f4
--- /dev/null
+++ b/vapi/Cogl-51.metadata
@@ -0,0 +1,40 @@
+* cheader_filename="cogl/cogl.h"
+
+Color struct
+
+Color.equal.v1 type="Cogl.Color"
+Color.equal.v2 type="Cogl.Color"
+Color.from_string name="init_from_string"
+
+Bitmap.* skip=false
+Bitmap.new_for_data.data type="owned uint8[]"
+Bitmap.get_buffer type="unowned Cogl.Buffer?"
+
+Primitive.new skip=false
+
+Texture
+  .get_data.data type="uint8[]"
+  .set_data.data type="uint8[]"
+  .set_region.data type="uint8[]"
+
+Texture2D
+  .new_from_data skip=false
+
+Texture.error_quark parent="Cogl.TextureError" name="quark"
+
+RendererError errordomain
+SystemError errordomain
+TextureError errordomain
+FramebufferError errordomain
+ScanoutError errordomain
+
+Onscreen
+  .add_frame_callback unowned
+
+A_BIT parent="Cogl.Bits" name="A"
+BGR_BIT parent="Cogl.Bits" name="BGR"
+AFIRST_BIT parent="Cogl.Bits" name="AFIRST"
+PREMULT_BIT parent="Cogl.Bits" name="PREMULT"
+DEPTH_BIT parent="Cogl.Bits" name="DEPTH"
+STENCIL_BIT parent="Cogl.Bits" name="STENCIL"
+PIXEL_FORMAT_MAX_PLANES parent = "Cogl.PixelFormat" name="MAX_PLANES"
diff --git a/vapi/Meta-51.metadata b/vapi/Meta-51.metadata
new file mode 100644
index 0000000..04355e7
--- /dev/null
+++ b/vapi/Meta-51.metadata
@@ -0,0 +1,176 @@
+* skip=false
+*.* skip=false
+* cheader_filename="meta/main.h"
+
+Backend cheader_filename="meta/meta-backend.h"
+Backend.gpu_added skip
+Background cheader_filename="meta/meta-background.h"
+Background.set_file.file nullable
+BackgroundContent.new symbol_type="constructor"
+BackgroundActor cheader_filename="meta/meta-background-actor.h"
+BackgroundContent cheader_filename="meta/meta-background-content.h"
+BackgroundGroup cheader_filename="meta/meta-background-group.h"
+BackgroundImage cheader_filename="meta/meta-background-image.h"
+BackgroundImageCache cheader_filename="meta/meta-background-image.h"
+Barrier cheader_filename="meta/barrier.h"
+BarrierDirection cheader_filename="meta/barrier.h"
+BarrierEvent cheader_filename="meta/barrier.h"
+BarrierFlags cheader_filename="meta/barrier.h"
+ButtonFunction cheader_filename="meta/common.h"
+ButtonLayout cheader_filename="meta/common.h"
+Compositor cheader_filename="meta/compositor.h"
+CompEffect cheader_filename="meta/compositor.h"
+CloseDialog cheader_filename="meta/meta-close-dialog.h"
+CloseDialogResponse cheader_filename="meta/meta-close-dialog.h"
+Context cheader_filename="meta/meta-context.h"
+CursorTracker cheader_filename="meta/meta-cursor-tracker.h"
+CursorTracker.get_pointer.mods out
+DebugControl cheader_filename="meta/meta-debug-control.h"
+DebugTopic cheader_filename="meta/util.h"
+DebugPaintFlag cheader_filename="meta/util.h"
+Direction cheader_filename="meta/common.h"
+Display cheader_filename="meta/display.h"
+Display.focus_window#signal name="do_focus_window"
+DisplayCorner cheader_filename="meta/display.h"
+DisplayDirection cheader_filename="meta/display.h"
+Dnd cheader_filename="meta/meta-dnd.h"
+EdgeType cheader_filename="meta/boxes.h"
+Edge cheader_filename="meta/boxes.h"
+FrameBorders cheader_filename="meta/common.h"
+FrameType cheader_filename="meta/common.h"
+GrabOp cheader_filename="meta/common.h"
+Gravity cheader_filename="meta/common.h"
+Group cheader_filename="meta/meta-x11-types.h"
+IdleMonitor cheader_filename="meta/meta-idle-monitor.h"
+IdleMonitorWatchFunc cheader_filename="meta/meta-idle-monitor.h"
+InhibitShortcutsDialog cheader_filename="meta/meta-inhibit-shortcuts-dialog.h"
+InhibitShortcutsDialogResponse cheader_filename="meta/meta-inhibit-shortcuts-dialog.h"
+KeyboardA11yFlags cheader_filename="meta/meta-enums.h"
+KeyBinding cheader_filename="meta/keybindings.h"
+keybindings_set_custom_handler parent="Meta.KeyBinding" name="set_custom_handler" cheader_filename="meta/keybindings.h"
+KeyBindingAction cheader_filename="meta/prefs.h"
+KeyBindingFlags cheader_filename="meta/prefs.h"
+KeyHandlerFunc cheader_filename="meta/prefs.h"
+KeyHandlerFunc.event type="Clutter.KeyEvent?"
+KeyHandlerFunc.window nullable
+LaunchContext cheader_filename="meta/meta-launch-context.h"
+Laters cheader_filename="meta/types.h"
+LaterType cheader_filename="meta/util.h"
+MaximizeFlags cheader_filename="meta/window.h"
+MultiTexture cheader_filename="meta/meta-multi-texture.h"
+MultiTextureFormat cheader_filename="meta/meta-multi-texture-format.h"
+MonitorManager cheader_filename="meta/meta-monitor-manager.h"
+MonitorSwitchConfigType cheader_filename="meta/meta-monitor-manager.h"
+MotionDirection cheader_filename="meta/common.h"
+Orientation cheader_filename="meta/meta-orientation-manager.h"
+OrientationManager cheader_filename="meta/meta-orientation-manager.h"
+PadDirection cheader_filename="meta/display.h"
+PadFeatureType cheader_filename="meta/display.h"
+Plugin cheader_filename="meta/meta-plugin.h"
+PowerSaveChangeReason cheader_filename="meta/meta-monitor-manager.h"
+Preference cheader_filename="meta/prefs.h"
+PrefsChangedFunc cheader_filename="meta/prefs.h"
+RemoteAccessController cheader_filename="meta/meta-remote-access-controller.h"
+RemoteAccessHandle cheader_filename="meta/meta-remote-access-controller.h"
+Selection cheader_filename="meta/meta-selection.h"
+SelectionSource cheader_filename="meta/meta-selection-source.h"
+SelectionSourceMemory cheader_filename="meta/meta-selection-source-memory.h"
+SelectionType cheader_filename="meta/meta-selection-source.h"
+Settings cheader_filename="meta/meta-settings.h"
+ShapedTexture cheader_filename="meta/meta-shaped-texture.h"
+Side cheader_filename="meta/common.h"
+SizeChange cheader_filename="meta/compositor.h"
+SoundPlayer cheader_filename="meta/meta-sound-player.h"
+StartupNotification cheader_filename="meta/meta-startup-notification.h"
+StartupNotification.changed.object type="Meta.StartupSequence"
+StartupNotification.get_sequences type_arguments="Meta.StartupSequence"
+StartupSequence cheader_filename="meta/meta-startup-notification.h"
+StackLayer cheader_filename="meta/common.h"
+Stage cheader_filename="meta/meta-stage.h"
+Strut cheader_filename="meta/boxes.h"
+TabList cheader_filename="meta/display.h"
+TabShowType cheader_filename="meta/display.h"
+WaylandClient cheader_filename="meta/meta-wayland-client.h"
+WaylandCompositor cheader_filename="meta/meta-wayland-compositor.h"
+Workspace cheader_filename="meta/workspace.h"
+WorkspaceManager cheader_filename="meta/meta-workspace-manager.h"
+Window cheader_filename="meta/window.h"
+Window.focus#signal name="focused"
+Window.suspend_state skip
+Window.main_monitor skip
+WindowActor cheader_filename="meta/meta-window-actor.h"
+WindowClientType cheader_filename="meta/window.h"
+WindowForeachFunc cheader_filename="meta/window.h"
+WindowGroup cheader_filename="meta/meta-window-group.h"
+WindowMenuType cheader_filename="meta/compositor.h"
+WindowType cheader_filename="meta/window.h"
+X11Display cheader_filename="meta/meta-x11-display.h"
+X11Display.add_event_func skip
+X11DisplayEventFunc skip
+
+WaylandSurface.scanout_candidate skip
+WaylandSurface.main_monitor skip
+
+// As per https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/2287
+MonitorManager.monitor_privacy_screen_changed skip
+
+prefs_* parent="Meta.Prefs" name="prefs_(.+)" cheader_filename="meta/prefs.h"
+
+g_utf8_strndup skip
+
+preference_to_string cheader_filename="meta/prefs.h"
+frame_type_to_string cheader_filename="meta/util.h"
+topic_to_string parent="Meta.DebugTopic" name="to_string" cheader_filename="meta/util.h"
+
+CURRENT_TIME cheader_filename="meta/common.h"
+PRIORITY_RESIZE cheader_filename="meta/common.h"
+PRIORITY_BEFORE_REDRAW cheader_filename="meta/common.h"
+PRIORITY_REDRAW cheader_filename="meta/common.h"
+PRIORITY_PREFS_NOTIFY cheader_filename="meta/common.h"
+VIRTUAL_CORE_POINTER_ID cheader_filename="meta/common.h"
+VIRTUAL_CORE_KEYBOARD_ID cheader_filename="meta/common.h"
+
+Display.window_visibility_updated
+	.object name="unplaced" type="GLib.List<unowned Meta.Window>"
+	.p0 name="should_show" type="GLib.List<unowned Meta.Window>"
+	.p1 name="should_hide" type="GLib.List<unowned Meta.Window>"
+
+Compositor.get_window_actors type_arguments="Meta.WindowActor"
+
+add_verbose_topic parent="Meta.Util" cheader_filename="meta/util.h"
+bug parent="Meta.Util" cheader_filename="meta/util.h"
+external_binding_name_for_action parent="Meta.Util" cheader_filename="meta/util.h"
+fatal parent="Meta.Util" cheader_filename="meta/util.h"
+is_verbose parent="Meta.Util" cheader_filename="meta/util.h"
+pop_no_msg_prefix parent="Meta.Util" cheader_filename="meta/util.h"
+push_no_msg_prefix parent="Meta.Util" cheader_filename="meta/util.h"
+remove_verbose_topic parent="Meta.Util" cheader_filename="meta/util.h"
+unsigned_long_equal parent="Meta.Util" name="ulong_equal" cheader_filename="meta/util.h"
+unsigned_long_equal.v1 type="ulong?"
+unsigned_long_equal.v2 type="ulong?"
+unsigned_long_hash parent="Meta.Util" name="ulong_hash" cheader_filename="meta/util.h"
+unsigned_long_hash.v type="ulong?"
+create_context parent="Meta.Context" name="new" symbol_type="constructor" cheader_filename="meta/meta-context.h"
+
+Plugin.create_close_dialog unowned=false nullable
+Plugin.create_inhibit_shortcuts_dialog unowned=false
+
+BackgroundActor sealed
+BackgroundContent sealed
+BackgroundImage sealed
+BackgroundImageCache sealed
+Background sealed
+Dnd sealed
+IdleMonitor sealed
+LaunchContext sealed
+RemoteAccessController sealed
+SelectionSourceMemory sealed
+Selection sealed
+ShapedTexture sealed
+SoundPlayer sealed
+Stage sealed
+StartupNotification sealed
+WaylandClient sealed
+WindowGroup sealed
+WorkspaceManager sealed
+X11Display sealed
diff --git a/vapi/Mtk-51.metadata b/vapi/Mtk-51.metadata
new file mode 100644
index 0000000..f804d1b
--- /dev/null
+++ b/vapi/Mtk-51.metadata
@@ -0,0 +1,6 @@
+Rectangle struct
+RECTANGLE_MAX_STACK_RECTS parent="Mtk.Rectangle" name="MAX_STACK_RECTS"
+REGION_BUILDER_MAX_LEVELS parent="Mtk.RegionBuilder" name="MAX_LEVELS"
+
+MONITOR_ALL_TRANSFORMS parent="Mtk.MonitorTransform" name="ALL"
+MONITOR_N_TRANSFORMS skip
diff --git a/vapi/libmutter-51.deps b/vapi/libmutter-51.deps
new file mode 120000
index 0000000..18c721d
--- /dev/null
+++ b/vapi/libmutter-51.deps
@@ -0,0 +1 @@
+libmutter-17.deps
\ No newline at end of file
diff --git a/vapi/libmutter-51.vapi b/vapi/libmutter-51.vapi
new file mode 120000
index 0000000..ff0e578
--- /dev/null
+++ b/vapi/libmutter-51.vapi
@@ -0,0 +1 @@
+libmutter-17.vapi
\ No newline at end of file
diff --git a/vapi/libmutter.vapi b/vapi/libmutter.vapi
index 922c695..6b1282b 100644
--- a/vapi/libmutter.vapi
+++ b/vapi/libmutter.vapi
@@ -29,8 +29,10 @@ namespace Meta {
 		public static bool get_center_new_windows ();
 		[CCode (cheader_filename = "meta/prefs.h")]
 		public static uint get_check_alive_timeout ();
+#if !HAS_MUTTER51
 		[CCode (cheader_filename = "meta/prefs.h")]
 		public static bool get_compositing_manager ();
+#endif
 		[CCode (cheader_filename = "meta/prefs.h")]
 		public static int get_cursor_size ();
 		[CCode (cheader_filename = "meta/prefs.h")]
@@ -1127,10 +1129,16 @@ namespace Meta {
 		public bool get_icon_geometry (out Mtk.Rectangle rect);
 		public uint64 get_id ();
 		public Meta.StackLayer get_layer ();
+#if HAS_MUTTER51
+		public bool get_max_size (out int width, out int height);
+#endif
 #if HAS_MUTTER49
 		public Meta.MaximizeFlags get_maximize_flags ();
 #else
 		public Meta.MaximizeFlags get_maximized ();
+#endif
+#if HAS_MUTTER51
+		public bool get_min_size (out int width, out int height);
 #endif
 		public int get_monitor ();
 		public unowned string? get_mutter_hints ();
diff --git a/vapi/meson.build b/vapi/meson.build
index 464346f..a62be81 100644
--- a/vapi/meson.build
+++ b/vapi/meson.build
@@ -483,3 +483,74 @@ if mutter50_dep.found()
 		output: 'libmutter-18.vapi'
 	)
 endif
+if mutter51_dep.found()
+	mtk_target = custom_target('mutter-mtk-51',
+		command: [
+			vapigen,
+			mutter_typelib_dir / 'Mtk-51.gir',
+			'--library=mutter-mtk-51',
+			'--pkg=gobject-2.0',
+			'--pkg=gio-2.0',
+			'--pkg=cairo',
+			'--pkg=graphene-gobject-1.0',
+			vapigen_args
+		],
+		output: 'mutter-mtk-51.vapi'
+	)
+	cogl_target = custom_target('mutter-cogl-51',
+		command: [
+			vapigen,
+			mutter_typelib_dir / 'Cogl-51.gir',
+			'--library=mutter-cogl-51',
+			'--pkg=mutter-mtk-51',
+			'--pkg=gobject-2.0',
+			'--pkg=gio-2.0',
+			'--pkg=cairo',
+			'--pkg=graphene-gobject-1.0',
+			'--pkg=x11',
+			vapigen_args,
+			files('Cogl-51-custom.vala')
+		],
+		depends: mtk_target,
+		output: 'mutter-cogl-51.vapi'
+	)
+
+	clutter_target = custom_target('mutter-clutter-51',
+		command: [
+			vapigen,
+			mutter_typelib_dir / 'Clutter-51.gir',
+			'--library=mutter-clutter-51',
+			'--pkg=graphene-gobject-1.0',
+			'--pkg=mutter-mtk-51',
+			'--pkg=mutter-cogl-51',
+			'--pkg=atk',
+			'--pkg=gio-2.0',
+			'--pkg=json-glib-1.0',
+			'--pkg=pango',
+			'--pkg=x11',
+			vapigen_args,
+		],
+		depends: [ cogl_target, mtk_target ],
+		output: 'mutter-clutter-51.vapi'
+	)
+
+	libmutter_target = custom_target('libmutter-51',
+		command: [
+			vapigen,
+			mutter_typelib_dir / 'Meta-51.gir',
+			'--library=libmutter-51',
+			'--pkg=graphene-gobject-1.0',
+			'--pkg=mutter-cogl-51',
+			'--pkg=mutter-clutter-51',
+			'--pkg=atk',
+			'--pkg=gio-2.0',
+			'--pkg=json-glib-1.0',
+			'--pkg=pango',
+			'--pkg=x11',
+			'--pkg=xfixes-4.0',
+			vapigen_args
+		],
+		depends: [ cogl_target, clutter_target ],
+		output: 'libmutter-51.vapi'
+	)
+endif
diff --git a/vapi/mutter-clutter-51.deps b/vapi/mutter-clutter-51.deps
new file mode 100644
index 0000000..9549905
--- /dev/null
+++ b/vapi/mutter-clutter-51.deps
@@ -0,0 +1,6 @@
+atk
+cairo
+pango
+json-glib-1.0
+mutter-cogl-51
+graphene-gobject-1.0
diff --git a/vapi/mutter-clutter-51.vapi b/vapi/mutter-clutter-51.vapi
new file mode 120000
index 0000000..ee388ea
--- /dev/null
+++ b/vapi/mutter-clutter-51.vapi
@@ -0,0 +1 @@
+mutter-clutter-17.vapi
\ No newline at end of file
diff --git a/vapi/mutter-clutter.vapi b/vapi/mutter-clutter.vapi
index ae1a0b0..ed89ea1 100644
--- a/vapi/mutter-clutter.vapi
+++ b/vapi/mutter-clutter.vapi
@@ -5219,6 +5219,10 @@ namespace Clutter {
 		public Cogl.Color get_background_color ();
 #else
 		public Clutter.Color get_background_color ();
+#endif
+#if HAS_MUTTER51
+		[CCode (cname = "clutter_actor_class_get_binding_pool")]
+		public class unowned Clutter.BindingPool get_binding_pool ();
 #endif
 		public unowned Clutter.Actor? get_child_at_index (int index_);
 		public Graphene.Matrix get_child_transform ();
@@ -6185,6 +6189,9 @@ namespace Clutter {
 		public double get_gesture_pinch_scale ();
 		public uint32 get_im_delete_length ();
 		public bool get_im_location (int32 offset, int32 anchor);
+#if HAS_MUTTER50
+		public bool get_im_preedit_hints (Clutter.PreeditAttribute preedit_hints, uint n_preedit_hints);
+#endif
 		public Clutter.PreeditResetMode get_im_preedit_reset_mode ();
 		public unowned string get_im_text ();
 		public uint16 get_key_code ();
@@ -6580,6 +6587,10 @@ namespace Clutter {
 	public abstract class InputFocus : GLib.Object {
 		[CCode (has_construct_function = false)]
 		protected InputFocus ();
+#if HAS_MUTTER51
+		[NoWrapper]
+		public virtual void action (Clutter.InputAction action);
+#endif
 		[NoWrapper]
 		public virtual void commit_text (string text);
 		[NoWrapper]
@@ -6600,7 +6611,11 @@ namespace Clutter {
 		public void set_cursor_location (Graphene.Rect rect);
 		public void set_input_panel_state (Clutter.InputPanelState state);
 		[NoWrapper]
+	#if HAS_MUTTER51
+		public virtual void set_preedit_text (string preedit, uint cursor, uint anchor, Clutter.PreeditAttribute style_hints, uint n_style_hints);
+	#else
 		public virtual void set_preedit_text (string preedit, uint cursor, uint anchor);
+	#endif
 		public void set_surrounding (string text, uint cursor, uint anchor);
 	}
 	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_input_method_get_type ()")]
@@ -6617,10 +6632,19 @@ namespace Clutter {
 		public virtual void reset ();
 		[NoWrapper]
 		public virtual void set_cursor_location (Graphene.Rect rect);
+#if HAS_MUTTER50
+		public virtual void set_handled_actions (Clutter.InputActionFlags actions);
+#endif
 		public void set_input_panel_state (Clutter.InputPanelState state);
 		public void set_preedit_text (string? preedit, uint cursor, uint anchor, Clutter.PreeditResetMode mode);
+#if HAS_MUTTER50
+		public void set_preedit_text_with_attrs (string? preedit, uint cursor, uint anchor, Clutter.PreeditResetMode mode, [CCode (array_length_cname = "n_preedit_hints", array_length_pos = 5.1, array_length_type = "guint")] Clutter.PreeditAttribute[]? preedit_hints);
+#endif
 		[NoWrapper]
 		public virtual void set_surrounding (string text, uint cursor, uint anchor);
+#if HAS_MUTTER50
+		public void trigger_action (Clutter.InputAction action);
+#endif
 		[NoWrapper]
 		public virtual void update_content_hints (Clutter.InputContentHintFlags hint);
 		[NoWrapper]
@@ -6673,6 +6697,23 @@ namespace Clutter {
 		public GLib.Value initial { owned get; set; }
 		public GLib.Type value_type { get; construct; }
 	}
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_key_controller_get_type ()")]
+	public sealed class KeyController : Clutter.Action {
+		[CCode (has_construct_function = false, type = "ClutterAction*")]
+		public KeyController (Clutter.InputFocus? im_focus);
+		public bool get_key (out uint32 key_symbol, out uint32 key_code, out unichar unicode);
+		public bool get_state (out Clutter.ModifierType pressed, out Clutter.ModifierType latched, out Clutter.ModifierType locked);
+		public bool get_trigger_keybindings ();
+		public void set_trigger_keybindings (bool trigger_keybindings);
+		[NoAccessorMethod]
+		public Clutter.InputFocus im_focus { construct; }
+		public bool trigger_keybindings { get; set; }
+		public signal bool key_press ();
+		public signal bool key_release ();
+		public signal void modifier_change ();
+	}
+#endif
 	[CCode (cheader_filename = "clutter/clutter.h", has_type_id = false)]
 	[Compact]
 	public class KeyEvent : Clutter.Event {
@@ -6803,6 +6844,16 @@ namespace Clutter {
 		[CCode (has_construct_function = false, type = "ClutterAction*")]
 		public LongPressGesture ();
 	}
+#endif
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_motion_controller_get_type ()")]
+	public class MotionController : Clutter.Action {
+		[CCode (has_construct_function = false, type = "ClutterAction*")]
+		public MotionController ();
+		public signal void enter (Clutter.Sprite sprite, float x, float y);
+		public signal void leave (Clutter.Sprite sprite);
+		public signal void motion (Clutter.Sprite sprite, float x, float y);
+	}
 #endif
 	[CCode (cheader_filename = "clutter/clutter.h", has_type_id = false)]
 	[Compact]
@@ -7154,6 +7205,16 @@ namespace Clutter {
 		public void set_scroll_mode (Clutter.ScrollMode mode);
 		public Clutter.ScrollMode scroll_mode { get; set; }
 	}
+#endif
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_scroll_controller_get_type ()")]
+	public class ScrollController : Clutter.Action {
+		[CCode (has_construct_function = false, type = "ClutterAction*")]
+		public ScrollController (Clutter.ScrollControllerFlags flags);
+		[NoAccessorMethod]
+		public Clutter.ScrollControllerFlags flags { get; construct; }
+		public signal void scroll (Clutter.Sprite sprite, Clutter.ScrollSource source, double dx, double dy);
+	}
 #endif
 	[CCode (cheader_filename = "clutter/clutter.h", has_type_id = false)]
 	[Compact]
@@ -7279,6 +7340,13 @@ namespace Clutter {
 #endif
 	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_shader_effect_get_type ()")]
 	public class ShaderEffect : Clutter.OffscreenEffect {
+#if HAS_MUTTER51
+		[CCode (has_construct_function = false)]
+		protected ShaderEffect ();
+		[NoWrapper]
+		public virtual Cogl.Snippet get_static_snippet ();
+		public void set_uniform_float (string name, int n_components, [CCode (array_length_cname = "total_count", array_length_pos = 2.5)] float[] value);
+#else
 		[CCode (has_construct_function = false, type = "ClutterEffect*")]
 #if HAS_MUTTER48
 		public ShaderEffect (Cogl.ShaderType shader_type);
@@ -7295,12 +7363,18 @@ namespace Clutter {
 		[NoWrapper]
 		public virtual string get_static_shader_source ();
 		public bool set_shader_source (string source);
+#endif
 		public void set_uniform_value (string name, GLib.Value value);
+#if HAS_MUTTER51
+		[CCode (has_construct_function = false, type = "ClutterEffect*")]
+		public ShaderEffect.with_snippet (Cogl.Snippet snippet);
+#else
 		[NoAccessorMethod]
 #if HAS_MUTTER48
 		public Cogl.ShaderType shader_type { construct; }
 #else
 		public Clutter.ShaderType shader_type { construct; }
+#endif
 #endif
 	}
 	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_shader_float_get_type ()")]
@@ -7398,6 +7472,9 @@ namespace Clutter {
 		public bool paint_to_buffer (Mtk.Rectangle rect, float scale, [CCode (array_length = false)] uint8[] data, int stride, Cogl.PixelFormat format, Clutter.PaintFlag paint_flags) throws GLib.Error;
 		public Clutter.Content paint_to_content (Mtk.Rectangle rect, float scale, Clutter.PaintFlag paint_flags) throws GLib.Error;
 		public void paint_to_framebuffer (Cogl.Framebuffer framebuffer, Mtk.Rectangle rect, float scale, Clutter.PaintFlag paint_flags);
+#if HAS_MUTTER51
+		public void paint_to_framebuffer_clipped (Cogl.Framebuffer framebuffer, Mtk.Rectangle rect, float scale, Clutter.ColorState color_state, Mtk.Region redraw_clip, Clutter.PaintFlag paint_flags);
+#endif
 #if !HAS_MUTTER49
 		public bool pointing_input_foreach (Clutter.StageInputForeachFunc func);
 #endif
@@ -7411,7 +7488,9 @@ namespace Clutter {
 		public void set_active (bool is_active);
 #endif
 		public void set_key_focus (Clutter.Actor? actor);
+#if !HAS_MUTTER51
 		public void set_minimum_size (uint width, uint height);
+#endif
 #if !HAS_MUTTER48
 		public void set_title (string title);
 #if HAS_MUTTER46
@@ -8233,6 +8312,14 @@ namespace Clutter {
 		public int dwell_delay;
 		public int dwell_threshold;
 	}
+#if HAS_MUTTER50
+	[CCode (cheader_filename = "clutter/clutter.h", has_type_id = false)]
+	public struct PreeditAttribute {
+		public Clutter.PreeditStyleHint hint;
+		public uint start;
+		public uint end;
+	}
+#endif
 #if HAS_MUTTER48
 	[CCode (cheader_filename = "clutter/clutter.h", has_type_id = false)]
 	public struct Primaries {
@@ -8619,6 +8706,9 @@ namespace Clutter {
 	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_PHASE_", type_id = "clutter_event_phase_get_type ()")]
 	public enum EventPhase {
 		CAPTURE,
+#if HAS_MUTTER51
+		TARGET,
+#endif
 		BUBBLE
 	}
 	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_", type_id = "clutter_event_type_get_type ()")]
@@ -8733,6 +8823,20 @@ namespace Clutter {
 		TOP,
 		BOTTOM
 	}
+#if HAS_MUTTER50
+	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_INPUT_ACTION_", type_id = "clutter_input_action_get_type ()")]
+	public enum InputAction {
+		SUBMIT,
+		LAST
+	}
+	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_INPUT_ACTION_FLAG_", type_id = "clutter_input_action_flags_get_type ()")]
+	[Flags]
+	public enum InputActionFlags {
+		NONE,
+		SUBMIT,
+		ALL
+	}
+#endif
 	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_INPUT_AXIS_", type_id = "clutter_input_axis_get_type ()")]
 	public enum InputAxis {
 		IGNORE,
@@ -8786,6 +8890,10 @@ namespace Clutter {
 		HIDDEN_TEXT,
 		SENSITIVE_DATA,
 		LATIN,
+#if HAS_MUTTER51
+		INHIBIT_OSK,
+		NO_EMOJI,
+#endif
 		MULTILINE
 	}
 	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_INPUT_CONTENT_PURPOSE_", type_id = "clutter_input_content_purpose_get_type ()")]
@@ -9019,6 +9127,19 @@ namespace Clutter {
 		CLEAR,
 		COMMIT
 	}
+#if HAS_MUTTER50
+	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_PREEDIT_STYLE_", type_id = "clutter_preedit_style_hint_get_type ()")]
+	public enum PreeditStyleHint {
+		NONE,
+		WHOLE,
+		SELECTION,
+		PREDICTION,
+		PREFIX,
+		SUFFIX,
+		SPELLING_ERROR,
+		COMPOSE_ERROR
+	}
+#endif
 	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_REPAINT_FLAGS_", type_id = "clutter_repaint_flags_get_type ()")]
 	[Flags]
 	public enum RepaintFlags {
@@ -9043,6 +9164,17 @@ namespace Clutter {
 		NEAREST,
 		TRILINEAR
 	}
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_SCROLL_CONTROLLER_FLAG_", type_id = "clutter_scroll_controller_flags_get_type ()")]
+	[Flags]
+	public enum ScrollControllerFlags {
+		NONE,
+		SCROLL_VERTICAL,
+		SCROLL_HORIZONTAL,
+		DISCRETE,
+		PHYSICAL_DIRECTION
+	}
+#endif
 	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_SCROLL_", type_id = "clutter_scroll_direction_get_type ()")]
 	public enum ScrollDirection {
 		UP,
diff --git a/vapi/mutter-cogl-14.vapi b/vapi/mutter-cogl-14.vapi
index c234e7b..2140aca 100644
--- a/vapi/mutter-cogl-14.vapi
+++ b/vapi/mutter-cogl-14.vapi
@@ -113,9 +113,17 @@ namespace Cogl {
 		public Cogl.BufferUpdateHint update_hint { set construct; }
 	}
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_context_get_type ()")]
+#if HAS_MUTTER51
+	public abstract class Context : GLib.Object {
+#else
 	public sealed class Context : GLib.Object {
+#endif
 		[CCode (has_construct_function = false)]
+#if HAS_MUTTER51
+		protected Context ();
+#else
 		public Context (Cogl.Display? display) throws GLib.Error;
+#endif
 #if HAS_MUTTER47
 		public void flush ();
 #else
@@ -159,14 +167,35 @@ namespace Cogl {
 		public bool is_hardware_accelerated ();
 #endif
 		public void set_named_pipeline (Cogl.PipelineKey key, Cogl.Pipeline? pipeline);
+#if HAS_MUTTER51
+		public void set_winsys_feature (Cogl.WinsysFeature feature, bool value);
+#endif
 #if !HAS_MUTTER50
 		public int64 timestamp_query_get_time_ns (Cogl.TimestampQuery query);
 #endif
+#if HAS_MUTTER51
+		public Cogl.Display display { get; construct; }
+#endif
+	}
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_context_egl_get_type ()")]
+	public class ContextEGL : Cogl.Context {
+		[CCode (has_construct_function = false, type = "CoglContext*")]
+		public ContextEGL (Cogl.Display display) throws GLib.Error;
 	}
+#endif
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_display_get_type ()")]
+#if HAS_MUTTER51
+	public abstract class Display : GLib.Object {
+#else
 	public sealed class Display : GLib.Object {
+#endif
 		[CCode (has_construct_function = false)]
-#if HAS_MUTTER47
+#if HAS_MUTTER51
+		protected Display ();
+		[NoWrapper]
+		public virtual void destroy ();
+#elif HAS_MUTTER47
 		public Display (Cogl.Renderer renderer);
 #else
 		public Display (Cogl.Renderer renderer, Cogl.OnscreenTemplate onscreen_template);
@@ -176,7 +205,21 @@ namespace Cogl {
 		public void set_onscreen_template (Cogl.OnscreenTemplate onscreen_template);
 #endif
 		public bool setup () throws GLib.Error;
+#if HAS_MUTTER51
+		public Cogl.Renderer renderer { get; construct; }
+#endif
 	}
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_display_egl_get_type ()")]
+	public class DisplayEGL : Cogl.Display {
+		[CCode (has_construct_function = false)]
+		public DisplayEGL (Cogl.Renderer renderer);
+		[NoWrapper]
+		public virtual void cleanup_context ();
+		[NoWrapper]
+		public virtual bool context_created () throws GLib.Error;
+	}
+#endif
 	[CCode (cheader_filename = "cogl/cogl.h", free_function = "cogl_dma_buf_handle_free", has_type_id = false)]
 	[Compact]
 	public class DmaBufHandle {
@@ -281,6 +324,9 @@ namespace Cogl {
 		[CCode (cheader_filename = "cogl/cogl.h", cname = "cogl_blit_framebuffer")]
 #endif
 		public bool blit (Cogl.Framebuffer dst, int src_x, int src_y, int dst_x, int dst_y, int width, int height) throws GLib.Error;
+#if HAS_MUTTER51
+		public bool blit_region (Cogl.Framebuffer dst, Mtk.Region src_region, int dst_x_offset, int dst_y_offset) throws GLib.Error;
+#endif
 #if !HAS_MUTTER47
 		public void cancel_fence_callback (Cogl.FenceClosure closure);
 #endif
@@ -477,7 +523,12 @@ namespace Cogl {
 		public void show ();
 #endif
 		public void swap_buffers (Cogl.FrameInfo frame_info, void* user_data);
-#if HAS_MUTTER48
+#if HAS_MUTTER51
+		[NoWrapper]
+		public virtual bool swap_buffers_with_damage (Mtk.Region region, Cogl.FrameInfo info);
+		[NoWrapper]
+		public virtual bool swap_region (Mtk.Region region, Cogl.FrameInfo info);
+#elif HAS_MUTTER48
 		public virtual void swap_buffers_with_damage (Mtk.Region region, Cogl.FrameInfo info);
 		public virtual void swap_region (Mtk.Region region, Cogl.FrameInfo info);
 #else
@@ -544,7 +595,9 @@ namespace Cogl {
 		public bool get_per_vertex_point_size ();
 		public float get_point_size ();
 		public int get_uniform_location (string uniform_name);
+#if !HAS_MUTTER51
 		public unowned Cogl.Program get_user_program ();
+#endif
 #if HAS_MUTTER48
 		public bool has_capability (GLib.Quark domain, uint capability);
 #endif
@@ -583,7 +636,9 @@ namespace Cogl {
 		public void set_uniform_int (int uniform_location, int n_components, int count, int value);
 		public void set_uniform_matrix (int uniform_location, int dimensions, int count, bool transpose, float value);
 #endif
+#if !HAS_MUTTER51
 		public void set_user_program (Cogl.Program program);
+#endif
 	}
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_pixel_buffer_get_type ()")]
 	public sealed class PixelBuffer : Cogl.Buffer {
@@ -640,6 +695,7 @@ namespace Cogl {
 		[CCode (has_construct_function = false)]
 		public Primitive.with_attributes (Cogl.VerticesMode mode, int n_vertices, [CCode (array_length_cname = "n_attributes", array_length_pos = 3.1)] Cogl.Attribute[] attributes);
 	}
+#if !HAS_MUTTER51
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_program_get_type ()")]
 	public sealed class Program : GLib.Object {
 		[CCode (has_construct_function = false)]
@@ -667,10 +723,17 @@ namespace Cogl {
 		[Version (deprecated = true, deprecated_since = "1.16")]
 		public void set_uniform_matrix (int uniform_location, int dimensions, bool transpose, [CCode (array_length_cname = "count", array_length_pos = 2.5)] float[] value);
 	}
+#endif
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_renderer_get_type ()")]
+#if HAS_MUTTER51
+	public abstract class Renderer : GLib.Object {
+		[CCode (has_construct_function = false)]
+		protected Renderer ();
+#else
 	public sealed class Renderer : GLib.Object {
 		[CCode (has_construct_function = false)]
 		public Renderer ();
+#endif
 #if !HAS_MUTTER47
 		public void add_constraint (Cogl.RendererConstraint constraint);
 #endif
@@ -691,6 +754,10 @@ namespace Cogl {
 #if HAS_MUTTER47
 		public void* get_proc_address (string name);
 #endif
+#if HAS_MUTTER51
+		public virtual int get_sync_fd ();
+#endif
+#if !HAS_MUTTER51
 #if HAS_MUTTER50
 		public unowned Cogl.Winsys get_winsys ();
 #elif HAS_MUTTER49
@@ -700,6 +767,7 @@ namespace Cogl {
 		public void* get_winsys_data ();
 #else
 		public Cogl.WinsysID get_winsys_id ();
+#endif
 #endif
 		public bool is_dma_buf_supported ();
 #if HAS_MUTTER48
@@ -714,6 +782,15 @@ namespace Cogl {
 #if !HAS_MUTTER47
 		public void set_winsys_id (Cogl.WinsysID winsys_id);
 #endif
+#if HAS_MUTTER51
+		[NoWrapper]
+		public virtual bool load_driver (Cogl.DriverId driver_id) throws GLib.Error;
+		public void set_driver_id (Cogl.DriverId driver);
+		[NoWrapper]
+		public virtual void update_sync ();
+		[NoAccessorMethod]
+		public Cogl.Driver driver { owned get; construct; }
+#else
 #if HAS_MUTTER50
 		public void set_custom_winsys (Cogl.Winsys winsys);
 #endif
@@ -725,7 +802,16 @@ namespace Cogl {
 		public void set_winsys (void* winsys);
 #endif
 #endif
+#endif
+	}
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_renderer_egl_get_type ()")]
+	public class RendererEGL : Cogl.Renderer {
+		[CCode (has_construct_function = false)]
+		public RendererEGL ();
+		public bool has_feature (Cogl.EGLWinsysFeature feature);
 	}
+#endif
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_scanout_get_type ()")]
 	public sealed class Scanout : GLib.Object {
 		[CCode (has_construct_function = false)]
@@ -749,6 +835,7 @@ namespace Cogl {
 		public void set_src_rect (Graphene.Rect rect);
 		public signal void scanout_failed (Cogl.Onscreen object);
 	}
+#if !HAS_MUTTER51
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_shader_get_type ()")]
 	public sealed class Shader : GLib.Object {
 		[CCode (has_construct_function = false)]
@@ -766,6 +853,7 @@ namespace Cogl {
 		[Version (deprecated = true, deprecated_since = "1.16")]
 		public void source (string source);
 	}
+#endif
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_snippet_get_type ()")]
 	public sealed class Snippet : GLib.Object {
 		[CCode (has_construct_function = false)]
@@ -876,6 +964,7 @@ namespace Cogl {
 		[CCode (has_construct_function = false, type = "CoglTexture*")]
 		public Texture2DSliced.with_size (Cogl.Context ctx, int width, int height, int max_waste);
 	}
+#if !HAS_MUTTER51
 #if HAS_MUTTER50
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_winsys_get_type ()")]
 	public abstract class Winsys : GLib.Object {
@@ -916,6 +1005,7 @@ namespace Cogl {
 	[Compact]
 	public class TimestampQuery {
 	}
+#endif
 #endif
 	[CCode (cheader_filename = "cogl/cogl.h", type_cname = "CoglScanoutBufferInterface", type_id = "cogl_scanout_buffer_get_type ()")]
 	public interface ScanoutBuffer : GLib.Object {
@@ -986,7 +1076,7 @@ namespace Cogl {
 		public void set_test_function (Cogl.DepthTestFunction function);
 		public void set_write_enabled (bool enable);
 	}
-#if HAS_MUTTER50
+#if HAS_MUTTER50 && !HAS_MUTTER51
 	[CCode (cheader_filename = "cogl/cogl.h", has_type_id = false)]
 	public struct DisplayEGL {
 		public void* platform;
@@ -1020,7 +1110,7 @@ namespace Cogl {
 		public Cogl.Color color;
 	}
 #endif
-#if HAS_MUTTER50
+#if HAS_MUTTER50 && !HAS_MUTTER51
 	[CCode (cheader_filename = "cogl/cogl.h", has_type_id = false)]
 	public struct RendererEGL {
 		public Cogl.EGLWinsysFeature private_features;
@@ -1237,8 +1327,10 @@ namespace Cogl {
 		OGL_FEATURE_ID_MAP_BUFFER_FOR_READ,
 		[CCode (cname = "COGL_FEATURE_ID_MAP_BUFFER_FOR_WRITE")]
 		OGL_FEATURE_ID_MAP_BUFFER_FOR_WRITE,
+#if !HAS_MUTTER51
 		[CCode (cname = "COGL_FEATURE_ID_FENCE")]
 		OGL_FEATURE_ID_FENCE,
+#endif
 		[CCode (cname = "COGL_FEATURE_ID_TEXTURE_RG")]
 		OGL_FEATURE_ID_TEXTURE_RG,
 		[CCode (cname = "COGL_FEATURE_ID_TEXTURE_RGBA1010102")]
@@ -1251,8 +1343,10 @@ namespace Cogl {
 		[CCode (cname = "COGL_FEATURE_ID_BUFFER_AGE")]
 		OGL_FEATURE_ID_BUFFER_AGE,
 #endif
+#if !HAS_MUTTER51
 		[CCode (cname = "COGL_FEATURE_ID_TEXTURE_EGL_IMAGE_EXTERNAL")]
 		OGL_FEATURE_ID_TEXTURE_EGL_IMAGE_EXTERNAL,
+#endif
 		[CCode (cname = "COGL_FEATURE_ID_BLIT_FRAMEBUFFER")]
 		OGL_FEATURE_ID_BLIT_FRAMEBUFFER,
 #if HAS_MUTTER47 && !HAS_MUTTER50
@@ -1464,6 +1558,7 @@ namespace Cogl {
 		EGL
 	}
 #endif
+#if !HAS_MUTTER51
 #if HAS_MUTTER48
 	[CCode (cheader_filename = "cogl/cogl.h", cprefix = "COGL_SHADER_TYPE_", type_id = "cogl_shader_type_get_type ()")]
 #else
@@ -1473,6 +1568,7 @@ namespace Cogl {
 		VERTEX,
 		FRAGMENT
 	}
+#endif
 	[CCode (cheader_filename = "cogl/cogl.h", cprefix = "COGL_SNIPPET_HOOK_", has_type_id = false)]
 	public enum SnippetHook {
 		VERTEX,
@@ -1560,7 +1656,11 @@ namespace Cogl {
 		SWAP_REGION_SYNCHRONIZED,
 #endif
 		BUFFER_AGE,
+#if HAS_MUTTER51
+		SYNC_FD,
+#else
 		SYNC_AND_COMPLETE_EVENT,
+#endif
 		N_FEATURES
 	}
 #if !HAS_MUTTER50
@@ -1623,8 +1723,15 @@ namespace Cogl {
 	[CCode (cheader_filename = "cogl/cogl.h", cprefix = "COGL_SYSTEM_ERROR_", has_type_id = false)]
 #endif
 	public errordomain SystemError {
+#if HAS_MUTTER51
+		[CCode (cname = "COGL_SYSTEM_ERROR_UNSUPPORTED")]
+		COGL_SYSTEM_ERROR_UNSUPPORTED,
+		[CCode (cname = "COGL_SYSTEM_ERROR_NO_MEMORY")]
+		COGL_SYSTEM_ERROR_NO_MEMORY
+#else
 		UNSUPPORTED,
 		NO_MEMORY
+#endif
 	}
 	[CCode (cheader_filename = "cogl/cogl.h", cprefix = "COGL_TEXTURE_ERROR_", has_type_id = false)]
 	public errordomain TextureError {
diff --git a/vapi/mutter-cogl-51.deps b/vapi/mutter-cogl-51.deps
new file mode 100644
index 0000000..35fbfd0
--- /dev/null
+++ b/vapi/mutter-cogl-51.deps
@@ -0,0 +1,5 @@
+pango
+glib-2.0
+gio-2.0
+mutter-mtk-51
+x11
diff --git a/vapi/mutter-cogl-51.vapi b/vapi/mutter-cogl-51.vapi
new file mode 120000
index 0000000..dca44a1
--- /dev/null
+++ b/vapi/mutter-cogl-51.vapi
@@ -0,0 +1 @@
+mutter-cogl-17.vapi
\ No newline at end of file
diff --git a/vapi/mutter-mtk-51.vapi b/vapi/mutter-mtk-51.vapi
new file mode 120000
index 0000000..efd075a
--- /dev/null
+++ b/vapi/mutter-mtk-51.vapi
@@ -0,0 +1 @@
+mutter-mtk-17.vapi
\ No newline at end of file
diff --git a/wingpanel-interface/meson.build b/wingpanel-interface/meson.build
index fcda54d..3b6d2b3 100644
--- a/wingpanel-interface/meson.build
+++ b/wingpanel-interface/meson.build
@@ -71,6 +71,17 @@ if mutter50_dep.found()
 	vala_flags = ['--define', 'HAS_MUTTER46', '--define', 'HAS_MUTTER47', '--define', 'HAS_MUTTER48', '--define', 'HAS_MUTTER49', '--define', 'HAS_MUTTER50']
 endif
 
+mutter51_dep = dependency('libmutter-51', version: ['>= 51', '< 52'], required: false)
+if mutter51_dep.found()
+	libmutter_dep = dependency('libmutter-51', version: '>= 51')
+	mutter_dep = [
+		libmutter_dep,
+		dependency('mutter-mtk-51'), dependency('mutter-cogl-51'),
+		dependency('mutter-clutter-51')
+	]
+	vala_flags = ['--define', 'HAS_MUTTER43', '--define', 'HAS_MUTTER44', '--define', 'HAS_MUTTER45', '--define', 'HAS_MUTTER46', '--define', 'HAS_MUTTER47', '--define', 'HAS_MUTTER48', '--define', 'HAS_MUTTER49', '--define', 'HAS_MUTTER50', '--define', 'HAS_MUTTER51']
+endif
+
 if mutter_dep.length() == 0
 	error ('No supported mutter library found!')
 endif
