diff --git a/data/gala.gschema.xml b/data/gala.gschema.xml
index ea079697..a726ae8b 100644
--- a/data/gala.gschema.xml
+++ b/data/gala.gschema.xml
@@ -204,6 +204,16 @@
       <summary>The shortcut to enable picture-in-picture window</summary>
       <description>The shortcut to show the selection area to choose a window.</description>
     </key>
+    <key type="as" name="brightness-brighter">
+      <default><![CDATA[['XF86MonBrightnessUp']]]></default>
+      <summary>Monitor brightness up</summary>
+      <description>Increase monitor brightness on supported panels</description>
+    </key>
+    <key type="as" name="brightness-dimmer">
+      <default><![CDATA[['XF86MonBrightnessDown']]]></default>
+      <summary>Monitor brightness down</summary>
+      <description>Decrease monitor brightness on supported panels</description>
+    </key>
   </schema>
 
   <schema path="/io/elementary/desktop/wm/animations/" id="io.elementary.desktop.wm.animations">
diff --git a/src/DBus.vala b/src/DBus.vala
index d02a7595..3bd4c926 100644
--- a/src/DBus.vala
+++ b/src/DBus.vala
@@ -19,6 +19,9 @@ public class Gala.DBus {
             (connection, name) => {
                 try {
                     connection.register_object ("/io/elementary/gala", WindowDragProvider.get_instance ());
+#if HAS_MUTTER50
+                    connection.register_object ("/io/elementary/gala/BrightnessManager", new BrightnessManager (wm.get_display ()));
+#endif
                 } catch (Error e) {
                     warning (e.message);
                 }
diff --git a/src/Misc/BrightnessManager.vala b/src/Misc/BrightnessManager.vala
new file mode 100644
index 00000000..ee1761cd
--- /dev/null
+++ b/src/Misc/BrightnessManager.vala
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2026 elementary, Inc. (https://elementary.io)
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ *
+ * Authored by: Denis Garaev <garaevdi@outlook.com>
+ */
+
+#if HAS_MUTTER50
+[DBus (name = "io.elementary.gala.BrightnessManager")]
+public class Gala.BrightnessManager : GLib.Object {
+    public int brightness {
+        get {
+            if (backlight == null) {
+                return -1;
+            }
+            int min, max;
+            int value;
+            backlight.get_brightness_info (out min, out max);
+            value = backlight.get_brightness ();
+            return ((value - min) * 100) / (max - min);
+        }
+        set {
+            if (brightness == value || backlight == null) {
+                return;
+            }
+            int min, max;
+            backlight.get_brightness_info (out min, out max);
+            backlight.set_brightness ((int) (value * (max - min) / 100) + min);
+        }
+    }
+
+    [DBus (visible = false)]
+    public unowned Meta.Display display { private get; construct; }
+    private unowned Meta.MonitorManager monitor_manager;
+    private unowned Meta.Backlight? backlight;
+
+    public BrightnessManager (Meta.Display display) {
+        Object (
+            display: display
+        );
+    }
+
+    construct {
+        monitor_manager = display.get_context ().get_backend ().get_monitor_manager ();
+        monitor_manager.monitors_changed.connect (update_available_backlights);
+
+        var keybinding_settings = new GLib.Settings ("io.elementary.desktop.wm.keybindings");
+        display.add_keybinding ("brightness-brighter", keybinding_settings, Meta.KeyBindingFlags.NONE, () => {
+            brightness += 5;
+        });
+        display.add_keybinding ("brightness-dimmer", keybinding_settings, Meta.KeyBindingFlags.NONE, () => {
+            brightness -= 5;
+        });
+
+        update_available_backlights ();
+    }
+
+    private void update_available_backlights () {
+        backlight = null;
+        unowned var monitors = monitor_manager.get_monitors ();
+        foreach (unowned var monitor in monitors) {
+            if (monitor.is_primary () && monitor.is_active ()) {
+                backlight = monitor.get_backlight ();
+            }
+        }
+    }
+}
+#endif
diff --git a/src/meson.build b/src/meson.build
index f4e36e4c..69deb72a 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -25,6 +25,7 @@ gala_bin_sources = files(
     'HotCorners/HotCorner.vala',
     'HotCorners/HotCornerManager.vala',
     'Misc/BlurManager.vala',
+    'Misc/BrightnessManager.vala',
     'Misc/DaemonManager.vala',
     'Misc/DBusAccelerator.vala',
     'Misc/DesktopIntegration.vala',
diff --git a/vapi/Meta-18.metadata b/vapi/Meta-18.metadata
index 05cf306e..d8525071 100644
--- a/vapi/Meta-18.metadata
+++ b/vapi/Meta-18.metadata
@@ -12,6 +12,7 @@ BackgroundContent cheader_filename="meta/meta-background-content.h"
 BackgroundGroup cheader_filename="meta/meta-background-group.h"
 BackgroundImage cheader_filename="meta/meta-background-image.h"
 BackgroundImageCache cheader_filename="meta/meta-background-image.h"
+Backlight cheader_filename="meta/meta-backlight.h"
 Barrier cheader_filename="meta/barrier.h"
 BarrierDirection cheader_filename="meta/barrier.h"
 BarrierEvent cheader_filename="meta/barrier.h"
@@ -56,6 +57,7 @@ KeyHandlerFunc.window nullable
 LaunchContext cheader_filename="meta/meta-launch-context.h"
 Laters cheader_filename="meta/types.h"
 LaterType cheader_filename="meta/util.h"
+LogicalMonitor cheader_filename="meta/meta-logical-monitor.h"
 MaximizeFlags cheader_filename="meta/window.h"
 MultiTexture cheader_filename="meta/meta-multi-texture.h"
 MultiTextureAlphaMode.n_meta_multi_texture_alpha_modes name="N_MODES"
@@ -67,6 +69,7 @@ MultiTextureCoefficients cheader_filename="meta/meta-multi-texture-format.h" cpr
 MultiTextureFormat.n_meta_multi_texture_formats name="N_FORMATS"
 MultiTextureFormat.meta_multi_texture_format_* name="meta_multi_texture_format_(.+)"
 MultiTextureFormat cheader_filename="meta/meta-multi-texture-format.h" cprefix="META_MULTI_TEXTURE_FORMAT_"
+Monitor cheader_filename="meta/meta-monitor.h"
 MonitorManager cheader_filename="meta/meta-monitor-manager.h"
 MonitorSwitchConfigType cheader_filename="meta/meta-monitor-manager.h"
 MotionDirection cheader_filename="meta/common.h"
diff --git a/vapi/libmutter.vapi b/vapi/libmutter.vapi
index 9b6a8157..ff133b5f 100644
--- a/vapi/libmutter.vapi
+++ b/vapi/libmutter.vapi
@@ -276,7 +276,7 @@ namespace Meta {
 		public void purge (GLib.File file);
 	}
 #if HAS_MUTTER49
-	[CCode (cheader_filename = "meta/main.h", type_id = "meta_backlight_get_type ()")]
+	[CCode (cheader_filename = "meta/meta-backlight.h", type_id = "meta_backlight_get_type ()")]
 	public abstract class Backlight : GLib.Object {
 		[CCode (has_construct_function = false)]
 		protected Backlight ();
@@ -647,14 +647,14 @@ namespace Meta {
 		public Meta.Workspace workspace { owned get; set; }
 	}
 #if HAS_MUTTER49
-	[CCode (cheader_filename = "meta/main.h", type_id = "meta_logical_monitor_get_type ()")]
+	[CCode (cheader_filename = "meta/meta-logical-monitor.h", type_id = "meta_logical_monitor_get_type ()")]
 	public class LogicalMonitor : GLib.Object {
 		[CCode (has_construct_function = false)]
 		protected LogicalMonitor ();
 		public unowned GLib.List<Meta.Monitor> get_monitors ();
 		public int get_number ();
 	}
-	[CCode (cheader_filename = "meta/main.h", type_id = "meta_monitor_get_type ()")]
+	[CCode (cheader_filename = "meta/meta-monitor.h", type_id = "meta_monitor_get_type ()")]
 	public class Monitor : GLib.Object {
 		[CCode (has_construct_function = false)]
 		protected Monitor ();
