diff --git a/lib/Gestures/ToucheggBackend.vala b/lib/Gestures/ToucheggBackend.vala
index c280efc0..b1a19fc9 100644
--- a/lib/Gestures/ToucheggBackend.vala
+++ b/lib/Gestures/ToucheggBackend.vala
@@ -64,7 +64,7 @@ private class Gala.ToucheggBackend : Object, GestureBackend {
     /**
      * Maximum number of reconnection attempts to the daemon.
      */
-    private const int MAX_RECONNECTION_ATTEMPTS = 10;
+    private const int MAX_RECONNECTION_ATTEMPTS = 1;
 
     /**
      * Time to sleep between reconnection attempts.
