From 8ec14ef867595e874363c02abc84c16366bf8d48 Mon Sep 17 00:00:00 2001
From: Denis Garaev <garaevdi@yandex.ru>
Date: Thu, 9 Jul 2026 13:32:43 +0300
Subject: [PATCH 1/5] Vapi: additions for mutter 50.3

---
 vapi/mutter-clutter.vapi | 47 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 47 insertions(+)

diff --git a/vapi/mutter-clutter.vapi b/vapi/mutter-clutter.vapi
index cf652deb0..ef263d976 100644
--- a/vapi/mutter-clutter.vapi
+++ b/vapi/mutter-clutter.vapi
@@ -6117,6 +6117,9 @@ namespace Clutter {
 		public double get_gesture_pinch_scale ();
 		public uint32 get_im_delete_length ();
 		public bool get_im_location (int32 offset, int32 anchor);
+#if HAS_MUTTER50
+		public bool get_im_preedit_hints (Clutter.PreeditAttribute preedit_hints, uint n_preedit_hints);
+#endif
 		public Clutter.PreeditResetMode get_im_preedit_reset_mode ();
 		public unowned string get_im_text ();
 		public uint16 get_key_code ();
@@ -6512,10 +6515,19 @@ namespace Clutter {
 		public virtual void reset ();
 		[NoWrapper]
 		public virtual void set_cursor_location (Graphene.Rect rect);
+#if HAS_MUTTER50
+		public virtual void set_handled_actions (Clutter.InputActionFlags actions);
+#endif
 		public void set_input_panel_state (Clutter.InputPanelState state);
 		public void set_preedit_text (string? preedit, uint cursor, uint anchor, Clutter.PreeditResetMode mode);
+#if HAS_MUTTER50
+		public void set_preedit_text_with_attrs (string? preedit, uint cursor, uint anchor, Clutter.PreeditResetMode mode, [CCode (array_length_cname = "n_preedit_hints", array_length_pos = 5.1, array_length_type = "guint")] Clutter.PreeditAttribute[]? preedit_hints);
+#endif
 		[NoWrapper]
 		public virtual void set_surrounding (string text, uint cursor, uint anchor);
+#if HAS_MUTTER50
+		public void trigger_action (Clutter.InputAction action);
+#endif
 		[NoWrapper]
 		public virtual void update_content_hints (Clutter.InputContentHintFlags hint);
 		[NoWrapper]
@@ -7857,6 +7869,14 @@ namespace Clutter {
 		public int dwell_delay;
 		public int dwell_threshold;
 	}
+#if HAS_MUTTER50
+	[CCode (cheader_filename = "clutter/clutter.h", has_type_id = false)]
+	public struct PreeditAttribute {
+		public Clutter.PreeditStyleHint hint;
+		public uint start;
+		public uint end;
+	}
+#endif
 #if HAS_MUTTER48
 	[CCode (cheader_filename = "clutter/clutter.h", has_type_id = false)]
 	public struct Primaries {
@@ -8302,6 +8322,20 @@ namespace Clutter {
 		TOP,
 		BOTTOM
 	}
+#if HAS_MUTTER50
+	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_INPUT_ACTION_", type_id = "clutter_input_action_get_type ()")]
+	public enum InputAction {
+		SUBMIT,
+		LAST
+	}
+	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_INPUT_ACTION_FLAG_", type_id = "clutter_input_action_flags_get_type ()")]
+	[Flags]
+	public enum InputActionFlags {
+		NONE,
+		SUBMIT,
+		ALL
+	}
+#endif
 	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_INPUT_AXIS_", type_id = "clutter_input_axis_get_type ()")]
 	public enum InputAxis {
 		IGNORE,
@@ -8575,6 +8609,19 @@ namespace Clutter {
 		CLEAR,
 		COMMIT
 	}
+#if HAS_MUTTER50
+	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_PREEDIT_STYLE_", type_id = "clutter_preedit_style_hint_get_type ()")]
+	public enum PreeditStyleHint {
+		NONE,
+		WHOLE,
+		SELECTION,
+		PREDICTION,
+		PREFIX,
+		SUFFIX,
+		SPELLING_ERROR,
+		COMPOSE_ERROR
+	}
+#endif
 	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_REPAINT_FLAGS_", type_id = "clutter_repaint_flags_get_type ()")]
 	[Flags]
 	public enum RepaintFlags {

From 2dcac9dee3d95352ba8d4d24b11e14ae7ecf2b41 Mon Sep 17 00:00:00 2001
From: Denis Garaev <garaevdi@yandex.ru>
Date: Sat, 11 Jul 2026 20:12:41 +0300
Subject: [PATCH 2/5] Vapi: update for mutter 51

---
 meson.build                 |  11 ++
 vapi/Clutter-51.metadata    | 146 +++++++++++++++++
 vapi/Cogl-51-custom.vala    |  66 ++++++++
 vapi/Cogl-51.metadata       |  40 +++++
 vapi/Meta-51-custom.vala    |   5 +
 vapi/Meta-51.metadata       | 184 ++++++++++++++++++++++
 vapi/Mtk-51.metadata        |   6 +
 vapi/libmutter-51.deps      |   1 +
 vapi/libmutter-51.vapi      |   1 +
 vapi/libmutter.vapi         |  60 +++++++
 vapi/meson.build            |  72 +++++++++
 vapi/mutter-clutter-51.deps |   6 +
 vapi/mutter-clutter-51.vapi |   1 +
 vapi/mutter-clutter.vapi    | 305 ++++++++++++++++++++++++++++++++++++
 vapi/mutter-cogl-14.vapi    | 131 +++++++++++++++-
 vapi/mutter-cogl-51.deps    |   5 +
 vapi/mutter-cogl-51.vapi    |   1 +
 vapi/mutter-mtk-14.vapi     |   2 +
 vapi/mutter-mtk-51.vapi     |   1 +
 19 files changed, 1040 insertions(+), 4 deletions(-)
 create mode 100644 vapi/Clutter-51.metadata
 create mode 100644 vapi/Cogl-51-custom.vala
 create mode 100644 vapi/Cogl-51.metadata
 create mode 100644 vapi/Meta-51-custom.vala
 create mode 100644 vapi/Meta-51.metadata
 create mode 100644 vapi/Mtk-51.metadata
 create mode 120000 vapi/libmutter-51.deps
 create mode 120000 vapi/libmutter-51.vapi
 create mode 100644 vapi/mutter-clutter-51.deps
 create mode 120000 vapi/mutter-clutter-51.vapi
 create mode 100644 vapi/mutter-cogl-51.deps
 create mode 120000 vapi/mutter-cogl-51.vapi
 create mode 120000 vapi/mutter-mtk-51.vapi

diff --git a/meson.build b/meson.build
index c37cb6091..e98acb039 100644
--- a/meson.build
+++ b/meson.build
@@ -140,6 +140,17 @@ if mutter50_dep.found()
 	vala_flags = ['--define', 'HAS_MUTTER47', '--define', 'HAS_MUTTER48', '--define', 'HAS_MUTTER49', '--define', 'HAS_MUTTER50']
 endif
 
+mutter51_dep = dependency('libmutter-51', version: ['>= 51', '< 52'], required: false)
+if mutter51_dep.found()
+	libmutter_dep = dependency('libmutter-51', version: '>= 51')
+	mutter_dep = [
+		libmutter_dep,
+		dependency('mutter-mtk-51'), dependency('mutter-cogl-51'),
+		dependency('mutter-clutter-51')
+	]
+	vala_flags = ['--define', 'HAS_MUTTER43', '--define', 'HAS_MUTTER44', '--define', 'HAS_MUTTER45', '--define', 'HAS_MUTTER46', '--define', 'HAS_MUTTER47', '--define', 'HAS_MUTTER48', '--define', 'HAS_MUTTER49', '--define', 'HAS_MUTTER50', '--define', 'HAS_MUTTER51']
+endif
+
 if mutter_dep.length() == 0
     error ('No supported mutter library found!')
 endif
diff --git a/vapi/Clutter-51.metadata b/vapi/Clutter-51.metadata
new file mode 100644
index 000000000..8c22eeeba
--- /dev/null
+++ b/vapi/Clutter-51.metadata
@@ -0,0 +1,146 @@
+// Non mini-object
+ActorBox struct
+Margin struct
+PaintVolume struct
+Perspective struct
+
+*.ref unowned
+* cheader_filename="clutter/clutter.h"
+
+// Fix the few clutter-pango headers
+Text cheader_filename="clutter/clutter-pango.h"
+TextBuffer cheader_filename="clutter/clutter-pango.h"
+TextNode cheader_filename="clutter/clutter-pango.h"
+Actor
+	.get_pango_context cheader_filename="clutter/clutter-pango.h"
+	.create_pango_context cheader_filename="clutter/clutter-pango.h"
+	.create_pango_layout cheader_filename="clutter/clutter-pango.h"
+
+
+Actor
+	.apply_transform.matrix ref
+	.get_abs_allocation_vertices.verts out=false
+Event.type#method name="get_type"
+
+// ???
+Actor.has_pointer#method name="get_has_pointer"
+
+// Not all backing symbols are deprecated
+Actor.pick deprecated=false
+
+// Nullable return values
+Actor
+	.get_parent nullable
+
+// The original CType has been overridden by the annotations
+Actor
+	.allocate.box ctype="const ClutterActorBox *"
+	.get_stage ctype="ClutterActor *"
+Action.handle_event.event ctype="const ClutterEvent *"
+
+// method/virtual-method/signal don't match
+Actor
+	.event#method name="emit_event"
+	.get_paint_volume#virtual_method name="get_paint_volume_vfunc"
+	.get_paint_volume#virtual_method.volume out
+Text
+	.activate#method name="try_activate"
+	.insert_text#signal skip
+TextBuffer.get_text#virtual_method name="get_text_with_length"
+
+// Default values
+Stage.read_pixels
+	.width default=-1
+	.height default=-1
+Stage.paint_to_buffer
+    .data type="uint8[]"
+Text
+	.position_to_coords.line_height default=null
+
+// Skipped by g-i for unknown reasons
+LayoutManager
+	.create_child_meta skip=false
+
+// We can use static strings
+PaintNode
+	.set_static_name skip=false
+
+// Variadic arguments
+Backend
+	.get_cogl_context skip=false
+Interval
+	.new skip=false
+	.get_interval skip=false
+	.set_final skip=false
+	.set_initial skip=false
+	.set_interval skip=false
+LayoutManager
+	.child_get skip=false
+	.child_set skip=false
+
+// Skipped upstream for unknown reasons
+Interval.register_progress_func skip=false
+
+// struct/class confusion
+ActorBox
+	.new skip
+	.from_vertices skip
+Margin
+	.new skip
+
+// Upstream
+Event
+	.get_position.position out
+
+FrameListenerIface skip
+FrameClock.new skip
+
+// Remove for clutter-2.0
+/////////////////////////
+
+StageView.layout skip
+
+Stage
+	.paint_view.redraw_clip type="Cairo.Region"
+
+// *Event should be compact classes derived from Clutter.Event
+Event.type skip=false
+AnyEvent struct=false base_type="Clutter.Event"
+ButtonEvent struct=false base_type="Clutter.Event"
+CrossingEvent struct=false base_type="Clutter.Event"
+DeviceEvent struct=false base_type="Clutter.Event"
+IMEvent struct=false base_type="Clutter.Event"
+KeyEvent struct=false base_type="Clutter.Event"
+MotionEvent struct=false base_type="Clutter.Event"
+PadButtonEvent struct=false base_type="Clutter.Event"
+PadDialEvent struct=false base_type="Clutter.Event"
+PadRingEvent struct=false base_type="Clutter.Event"
+PadStripEvent struct=false base_type="Clutter.Event"
+ProximityEvent struct=false base_type="Clutter.Event"
+ScrollEvent struct=false base_type="Clutter.Event"
+TouchEvent struct=false base_type="Clutter.Event"
+TouchpadHoldEvent struct=false base_type="Clutter.Event"
+TouchpadPinchEvent struct=false base_type="Clutter.Event"
+TouchpadSwipeEvent struct=false base_type="Clutter.Event"
+
+// Keysyms used to be CLUTTER_X instead of CLUTTER_KEY_X
+*#constant skip
+CURRENT_TIME skip=false
+PRIORITY_REDRAW skip=false
+
+// Clutter devs don't like us creating nested namespaces
+value_* name="value_(.+)" parent="Clutter.Value"
+threads_* name="threads_(.+)" parent="Clutter.Threads"
+
+// There is no way to know sealed classes before GLib 2.70
+ColorState sealed
+FrameClock sealed
+TextureContent sealed
+
+TextureContent.new_from_texture symbol_type="constructor"
+
+// Possibly keep
+KEY_* skip=false name="KEY_(.+)" type="uint" parent="Clutter.Key"
+BUTTON_* skip=false name="BUTTON_(.+)" type="uint32" parent="Clutter.Button"
+EVENT_STOP skip=false type="bool"
+EVENT_PROPAGATE skip=false type="bool"
diff --git a/vapi/Cogl-51-custom.vala b/vapi/Cogl-51-custom.vala
new file mode 100644
index 000000000..90d46b330
--- /dev/null
+++ b/vapi/Cogl-51-custom.vala
@@ -0,0 +1,66 @@
+namespace Cogl {
+	public struct Color {
+		[CCode (cname="cogl_color_init_from_4f")]
+		public Color.from_4f (float red, float green, float blue, float alpha);
+		[CCode (cname="cogl_color_init_from_hsl")]
+		public Color.from_hsl (float hue, float saturation, float luminance);
+		[CCode (cname = "_vala_cogl_color_from_string")]
+		public static Cogl.Color? from_string (string str) {
+			Cogl.Color color = {};
+			if (color.init_from_string (str))
+				return color;
+
+			return null;
+		}
+	}
+	[CCode (cheader_filename = "cogl/cogl.h", has_type_id = false)]
+	public struct VertexP2 {
+		public float x;
+		public float y;
+	}
+	[CCode (cheader_filename = "cogl/cogl.h", has_type_id = false)]
+	public struct VertexP2C4 {
+		public float x;
+		public float y;
+		public uint8 r;
+		public uint8 g;
+		public uint8 b;
+		public uint8 a;
+	}
+	[CCode (cheader_filename = "cogl/cogl.h", has_type_id = false)]
+	public struct VertexP2T2 {
+		public float x;
+		public float y;
+		public float s;
+		public float t;
+	}
+	[CCode (cheader_filename = "cogl/cogl.h", has_type_id = false)]
+	public struct VertexP2T2C4 {
+		public float x;
+		public float y;
+		public float s;
+		public float t;
+		public uint8 r;
+		public uint8 g;
+		public uint8 b;
+		public uint8 a;
+	}
+	[CCode (cheader_filename = "cogl/cogl.h", has_type_id = false)]
+	public struct VertexP3 {
+		public float x;
+		public float y;
+		public float z;
+	}
+	[CCode (cheader_filename = "cogl/cogl.h", has_type_id = false)]
+	public struct VertexP3T2 {
+		public float x;
+		public float y;
+		public float z;
+		public float s;
+		public float t;
+	}
+	[CCode (cheader_filename = "cogl/cogl.h", cprefix = "COGL_PIXEL_FORMAT_", type_id = "cogl_pixel_format_get_type ()")]
+	public enum PixelFormat {
+		CAIRO_ARGB32_COMPAT;
+	}
+}
diff --git a/vapi/Cogl-51.metadata b/vapi/Cogl-51.metadata
new file mode 100644
index 000000000..f66d6f40f
--- /dev/null
+++ b/vapi/Cogl-51.metadata
@@ -0,0 +1,40 @@
+* cheader_filename="cogl/cogl.h"
+
+Color struct
+
+Color.equal.v1 type="Cogl.Color"
+Color.equal.v2 type="Cogl.Color"
+Color.from_string name="init_from_string"
+
+Bitmap.* skip=false
+Bitmap.new_for_data.data type="owned uint8[]"
+Bitmap.get_buffer type="unowned Cogl.Buffer?"
+
+Primitive.new skip=false
+
+Texture
+  .get_data.data type="uint8[]"
+  .set_data.data type="uint8[]"
+  .set_region.data type="uint8[]"
+
+Texture2D
+  .new_from_data skip=false
+
+Texture.error_quark parent="Cogl.TextureError" name="quark"
+
+RendererError errordomain
+SystemError errordomain
+TextureError errordomain
+FramebufferError errordomain
+ScanoutError errordomain
+
+Onscreen
+  .add_frame_callback unowned
+
+A_BIT parent="Cogl.Bits" name="A"
+BGR_BIT parent="Cogl.Bits" name="BGR"
+AFIRST_BIT parent="Cogl.Bits" name="AFIRST"
+PREMULT_BIT parent="Cogl.Bits" name="PREMULT"
+DEPTH_BIT parent="Cogl.Bits" name="DEPTH"
+STENCIL_BIT parent="Cogl.Bits" name="STENCIL"
+PIXEL_FORMAT_MAX_PLANES parent = "Cogl.PixelFormat" name="MAX_PLANES"
diff --git a/vapi/Meta-51-custom.vala b/vapi/Meta-51-custom.vala
new file mode 100644
index 000000000..6c2ad8938
--- /dev/null
+++ b/vapi/Meta-51-custom.vala
@@ -0,0 +1,5 @@
+namespace Meta.Util {
+    public static inline bool is_wayland_compositor () {
+        return true;
+    }
+}
diff --git a/vapi/Meta-51.metadata b/vapi/Meta-51.metadata
new file mode 100644
index 000000000..727fc2f68
--- /dev/null
+++ b/vapi/Meta-51.metadata
@@ -0,0 +1,184 @@
+* skip=false
+*.* skip=false
+* cheader_filename="meta/main.h"
+
+Backend cheader_filename="meta/meta-backend.h"
+Backend.gpu_added skip
+Background cheader_filename="meta/meta-background.h"
+Background.set_file.file nullable
+BackgroundContent.new symbol_type="constructor"
+BackgroundActor cheader_filename="meta/meta-background-actor.h"
+BackgroundContent cheader_filename="meta/meta-background-content.h"
+BackgroundGroup cheader_filename="meta/meta-background-group.h"
+BackgroundImage cheader_filename="meta/meta-background-image.h"
+BackgroundImageCache cheader_filename="meta/meta-background-image.h"
+Barrier cheader_filename="meta/barrier.h"
+BarrierDirection cheader_filename="meta/barrier.h"
+BarrierEvent cheader_filename="meta/barrier.h"
+BarrierFlags cheader_filename="meta/barrier.h"
+ButtonFunction cheader_filename="meta/common.h"
+ButtonLayout cheader_filename="meta/common.h"
+Compositor cheader_filename="meta/compositor.h"
+CompEffect cheader_filename="meta/compositor.h"
+CloseDialog cheader_filename="meta/meta-close-dialog.h"
+CloseDialogResponse cheader_filename="meta/meta-close-dialog.h"
+Context cheader_filename="meta/meta-context.h"
+Cursor cheader_filename="meta/meta-cursor.h"
+CursorTracker cheader_filename="meta/meta-cursor-tracker.h"
+CursorTracker.get_pointer.mods out
+DebugControl cheader_filename="meta/meta-debug-control.h"
+DebugTopic cheader_filename="meta/util.h"
+DebugPaintFlag cheader_filename="meta/util.h"
+Direction cheader_filename="meta/common.h"
+Display cheader_filename="meta/display.h"
+Display.focus_window#signal name="do_focus_window"
+DisplayCorner cheader_filename="meta/display.h"
+DisplayDirection cheader_filename="meta/display.h"
+Dnd cheader_filename="meta/meta-dnd.h"
+EdgeType cheader_filename="meta/boxes.h"
+Edge cheader_filename="meta/boxes.h"
+FrameBorders cheader_filename="meta/common.h"
+FrameType cheader_filename="meta/common.h"
+GrabOp cheader_filename="meta/common.h"
+Gravity cheader_filename="meta/common.h"
+Group cheader_filename="meta/meta-x11-types.h"
+IdleMonitor cheader_filename="meta/meta-idle-monitor.h"
+IdleMonitorWatchFunc cheader_filename="meta/meta-idle-monitor.h"
+InhibitShortcutsDialog cheader_filename="meta/meta-inhibit-shortcuts-dialog.h"
+InhibitShortcutsDialogResponse cheader_filename="meta/meta-inhibit-shortcuts-dialog.h"
+KeyboardA11yFlags cheader_filename="meta/meta-enums.h"
+KeyBinding cheader_filename="meta/keybindings.h"
+keybindings_set_custom_handler parent="Meta.KeyBinding" name="set_custom_handler" cheader_filename="meta/keybindings.h"
+KeyBindingAction cheader_filename="meta/prefs.h"
+KeyBindingFlags cheader_filename="meta/prefs.h"
+KeyHandlerFunc cheader_filename="meta/prefs.h"
+KeyHandlerFunc.event type="Clutter.KeyEvent?"
+KeyHandlerFunc.window nullable
+LaunchContext cheader_filename="meta/meta-launch-context.h"
+Laters cheader_filename="meta/types.h"
+LaterType cheader_filename="meta/util.h"
+MaximizeFlags cheader_filename="meta/window.h"
+MultiTexture cheader_filename="meta/meta-multi-texture.h"
+MultiTextureAlphaMode.n_meta_multi_texture_alpha_modes name="N_MODES"
+MultiTextureAlphaMode.meta_multi_texture_alpha_mode_* name="meta_multi_texture_alpha_mode_(.+)"
+MultiTextureAlphaMode cheader_filename="meta/meta-multi-texture-format.h" cprefix="META_MULTI_TEXTURE_ALPHA_MODE_"
+MultiTextureCoefficients.n_meta_multi_texture_coefficients name="N_COEFFICIENTS"
+MultiTextureCoefficients.meta_multi_texture_coefficients_* name="meta_multi_texture_coefficients_(.+)"
+MultiTextureCoefficients cheader_filename="meta/meta-multi-texture-format.h" cprefix="META_MULTI_TEXTURE_COEFFICIENTS_"
+MultiTextureFormat.n_meta_multi_texture_formats name="N_FORMATS"
+MultiTextureFormat.meta_multi_texture_format_* name="meta_multi_texture_format_(.+)"
+MultiTextureFormat cheader_filename="meta/meta-multi-texture-format.h" cprefix="META_MULTI_TEXTURE_FORMAT_"
+LogicalMonitor cheader_filename="meta/meta-logical-monitor.h"
+Monitor cheader_filename="meta/meta-monitor.h"
+MonitorManager cheader_filename="meta/meta-monitor-manager.h"
+MonitorSwitchConfigType cheader_filename="meta/meta-monitor-manager.h"
+MotionDirection cheader_filename="meta/common.h"
+Orientation cheader_filename="meta/meta-orientation-manager.h"
+OrientationManager cheader_filename="meta/meta-orientation-manager.h"
+PadDirection cheader_filename="meta/display.h"
+PadFeatureType cheader_filename="meta/display.h"
+Plugin cheader_filename="meta/meta-plugin.h"
+PowerSaveChangeReason cheader_filename="meta/meta-monitor-manager.h"
+Preference cheader_filename="meta/prefs.h"
+PrefsChangedFunc cheader_filename="meta/prefs.h"
+RemoteAccessController cheader_filename="meta/meta-remote-access-controller.h"
+RemoteAccessHandle cheader_filename="meta/meta-remote-access-controller.h"
+Selection cheader_filename="meta/meta-selection.h"
+SelectionSource cheader_filename="meta/meta-selection-source.h"
+SelectionSourceMemory cheader_filename="meta/meta-selection-source-memory.h"
+SelectionType cheader_filename="meta/meta-selection-source.h"
+Settings cheader_filename="meta/meta-settings.h"
+ShapedTexture cheader_filename="meta/meta-shaped-texture.h"
+Side cheader_filename="meta/common.h"
+SizeChange cheader_filename="meta/compositor.h"
+SoundPlayer cheader_filename="meta/meta-sound-player.h"
+StartupNotification cheader_filename="meta/meta-startup-notification.h"
+StartupNotification.changed.object type="Meta.StartupSequence"
+StartupNotification.get_sequences type_arguments="Meta.StartupSequence"
+StartupSequence cheader_filename="meta/meta-startup-notification.h"
+StackLayer cheader_filename="meta/common.h"
+Stage cheader_filename="meta/meta-stage.h"
+Strut cheader_filename="meta/boxes.h"
+TabList cheader_filename="meta/display.h"
+TabShowType cheader_filename="meta/display.h"
+WaylandClient cheader_filename="meta/meta-wayland-client.h"
+WaylandCompositor cheader_filename="meta/meta-wayland-compositor.h"
+Workspace cheader_filename="meta/workspace.h"
+WorkspaceManager cheader_filename="meta/meta-workspace-manager.h"
+Window cheader_filename="meta/window.h"
+Window.focus#signal name="focused"
+Window.suspend_state skip
+Window.main_monitor skip
+WindowActor cheader_filename="meta/meta-window-actor.h"
+WindowClientType cheader_filename="meta/window.h"
+WindowForeachFunc cheader_filename="meta/window.h"
+WindowGroup cheader_filename="meta/meta-window-group.h"
+WindowMenuType cheader_filename="meta/compositor.h"
+WindowType cheader_filename="meta/window.h"
+X11Display cheader_filename="meta/meta-x11-display.h"
+X11Display.add_event_func skip
+X11DisplayEventFunc skip
+
+WaylandSurface.scanout_candidate skip
+WaylandSurface.main_monitor skip
+
+// As per https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/2287
+MonitorManager.monitor_privacy_screen_changed skip
+
+prefs_* parent="Meta.Prefs" name="prefs_(.+)" cheader_filename="meta/prefs.h"
+
+g_utf8_strndup skip
+
+preference_to_string cheader_filename="meta/prefs.h"
+frame_type_to_string cheader_filename="meta/util.h"
+topic_to_string parent="Meta.DebugTopic" name="to_string" cheader_filename="meta/util.h"
+
+CURRENT_TIME cheader_filename="meta/common.h"
+PRIORITY_RESIZE cheader_filename="meta/common.h"
+PRIORITY_BEFORE_REDRAW cheader_filename="meta/common.h"
+PRIORITY_REDRAW cheader_filename="meta/common.h"
+PRIORITY_PREFS_NOTIFY cheader_filename="meta/common.h"
+VIRTUAL_CORE_POINTER_ID cheader_filename="meta/common.h"
+VIRTUAL_CORE_KEYBOARD_ID cheader_filename="meta/common.h"
+
+Display.window_visibility_updated
+	.object name="unplaced" type="GLib.List<unowned Meta.Window>"
+	.p0 name="should_show" type="GLib.List<unowned Meta.Window>"
+	.p1 name="should_hide" type="GLib.List<unowned Meta.Window>"
+
+Compositor.get_window_actors type_arguments="Meta.WindowActor"
+
+add_verbose_topic parent="Meta.Util" cheader_filename="meta/util.h"
+bug parent="Meta.Util" cheader_filename="meta/util.h"
+external_binding_name_for_action parent="Meta.Util" cheader_filename="meta/util.h"
+fatal parent="Meta.Util" cheader_filename="meta/util.h"
+is_verbose parent="Meta.Util" cheader_filename="meta/util.h"
+remove_verbose_topic parent="Meta.Util" cheader_filename="meta/util.h"
+unsigned_long_equal parent="Meta.Util" name="ulong_equal" cheader_filename="meta/util.h"
+unsigned_long_equal.v1 type="ulong?"
+unsigned_long_equal.v2 type="ulong?"
+unsigned_long_hash parent="Meta.Util" name="ulong_hash" cheader_filename="meta/util.h"
+unsigned_long_hash.v type="ulong?"
+create_context parent="Meta.Context" name="new" symbol_type="constructor" cheader_filename="meta/meta-context.h"
+
+Plugin.create_close_dialog unowned=false nullable
+Plugin.create_inhibit_shortcuts_dialog unowned=false
+
+BackgroundActor sealed
+BackgroundContent sealed
+BackgroundImage sealed
+BackgroundImageCache sealed
+Background sealed
+Dnd sealed
+IdleMonitor sealed
+LaunchContext sealed
+RemoteAccessController sealed
+SelectionSourceMemory sealed
+Selection sealed
+ShapedTexture sealed
+SoundPlayer sealed
+Stage sealed
+StartupNotification sealed
+WaylandClient sealed
+WindowGroup sealed
+WorkspaceManager sealed
diff --git a/vapi/Mtk-51.metadata b/vapi/Mtk-51.metadata
new file mode 100644
index 000000000..f804d1bc0
--- /dev/null
+++ b/vapi/Mtk-51.metadata
@@ -0,0 +1,6 @@
+Rectangle struct
+RECTANGLE_MAX_STACK_RECTS parent="Mtk.Rectangle" name="MAX_STACK_RECTS"
+REGION_BUILDER_MAX_LEVELS parent="Mtk.RegionBuilder" name="MAX_LEVELS"
+
+MONITOR_ALL_TRANSFORMS parent="Mtk.MonitorTransform" name="ALL"
+MONITOR_N_TRANSFORMS skip
diff --git a/vapi/libmutter-51.deps b/vapi/libmutter-51.deps
new file mode 120000
index 000000000..5209af8bf
--- /dev/null
+++ b/vapi/libmutter-51.deps
@@ -0,0 +1 @@
+libmutter-18.deps
\ No newline at end of file
diff --git a/vapi/libmutter-51.vapi b/vapi/libmutter-51.vapi
new file mode 120000
index 000000000..d08c01a0c
--- /dev/null
+++ b/vapi/libmutter-51.vapi
@@ -0,0 +1 @@
+libmutter-18.vapi
\ No newline at end of file
diff --git a/vapi/libmutter.vapi b/vapi/libmutter.vapi
index 9b6a81576..79e3c70a4 100644
--- a/vapi/libmutter.vapi
+++ b/vapi/libmutter.vapi
@@ -29,8 +29,10 @@ namespace Meta {
 		public static bool get_center_new_windows ();
 		[CCode (cheader_filename = "meta/prefs.h")]
 		public static uint get_check_alive_timeout ();
+#if !HAS_MUTTER51
 		[CCode (cheader_filename = "meta/prefs.h")]
 		public static bool get_compositing_manager ();
+#endif
 		[CCode (cheader_filename = "meta/prefs.h")]
 		public static int get_cursor_size ();
 		[CCode (cheader_filename = "meta/prefs.h")]
@@ -129,7 +131,9 @@ namespace Meta {
 	public abstract class Backend : GLib.Object, GLib.Initable {
 		[CCode (has_construct_function = false)]
 		protected Backend ();
+#if !HAS_MUTTER51
 		public void freeze_keyboard (uint32 timestamp);
+#endif
 		public Meta.BackendCapabilities get_capabilities ();
 		public unowned Meta.Context get_context ();
 		public unowned Meta.IdleMonitor get_core_idle_monitor ();
@@ -176,6 +180,9 @@ namespace Meta {
 #endif
 		public Meta.BackendCapabilities capabilities { get; }
 		public Meta.Context context { get; construct; }
+#if HAS_MUTTER51
+		public Clutter.InputDevice last_input_device { get; }
+#endif
 		public signal void keymap_changed ();
 		public signal void keymap_layout_group_changed (uint object);
 		public signal void last_device_changed (Clutter.InputDevice object);
@@ -342,6 +349,9 @@ namespace Meta {
 		public unowned Meta.Backend get_backend ();
 #endif
 		public unowned Clutter.Actor get_feedback_group ();
+#if HAS_MUTTER51
+		public unowned Clutter.Actor get_input_panel_group ();
+#endif
 		public unowned Meta.Laters get_laters ();
 #if HAS_MUTTER48
 		public unowned Clutter.Stage get_stage ();
@@ -398,6 +408,22 @@ namespace Meta {
 		public signal void prepare_shutdown ();
 		public signal void started ();
 	}
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "meta/meta-cursor.h", type_id = "meta_cursor_get_type ()")]
+	public abstract class Cursor : Clutter.Cursor {
+		[CCode (has_construct_function = false)]
+		protected Cursor ();
+		public unowned Meta.Backend get_backend ();
+		public uint get_size ();
+		public unowned string get_theme_name ();
+		[NoAccessorMethod]
+		public Meta.Backend backend { construct; }
+		[NoAccessorMethod]
+		public uint size { construct; }
+		[NoAccessorMethod]
+		public string theme_name { construct; }
+	}
+#endif
 	[CCode (cheader_filename = "meta/meta-cursor-tracker.h", type_id = "meta_cursor_tracker_get_type ()")]
 	public class CursorTracker : GLib.Object {
 		[CCode (has_construct_function = false)]
@@ -520,7 +546,11 @@ namespace Meta {
 		[CCode (cname = "focus-window")]
 		public signal void do_focus_window (Meta.Window object, int64 p0);
 		public signal void gl_video_memory_purged ();
+#if HAS_MUTTER51
+		public signal void grab_op_begin (Meta.Window object, Meta.GrabOp p0, Clutter.Sprite p1);
+#else
 		public signal void grab_op_begin (Meta.Window object, Meta.GrabOp p0);
+#endif
 		public signal void grab_op_end (Meta.Window object, Meta.GrabOp p0);
 		public signal void in_fullscreen_changed ();
 		public signal bool init_xserver (GLib.Task object);
@@ -659,12 +689,18 @@ namespace Meta {
 		[CCode (has_construct_function = false)]
 		protected Monitor ();
 		public unowned Meta.Backlight? get_backlight ();
+#if HAS_MUTTER51
+		public unowned string get_color_mode_string ();
+#endif
 		public unowned string get_connector ();
 		public unowned string get_display_name ();
 		public unowned string get_product ();
 		public unowned string get_serial ();
 		public unowned string get_vendor ();
 		public bool is_active ();
+#if HAS_MUTTER51
+		public bool is_available ();
+#endif
 		public bool is_builtin ();
 		public bool is_primary ();
 		public bool is_virtual ();
@@ -889,8 +925,13 @@ namespace Meta {
 	public sealed class ShapedTexture : GLib.Object, Clutter.Content {
 		[CCode (has_construct_function = false)]
 		protected ShapedTexture ();
+#if !HAS_MUTTER
 		public Cairo.Surface? get_image (Mtk.Rectangle? clip);
+#endif
 		public unowned Meta.MultiTexture get_texture ();
+#if HAS_MUTTER51
+		public Cogl.Bitmap? paint_to_bitmap (Mtk.Rectangle? clip, Cogl.PixelFormat format);
+#endif
 		public void set_create_mipmaps (bool create_mipmaps);
 		public void set_mask_texture (Cogl.Texture mask_texture);
 #if HAS_MUTTER47
@@ -1044,16 +1085,24 @@ namespace Meta {
 		public unowned string? get_gtk_application_id ();
 		public unowned string? get_gtk_application_object_path ();
 		public unowned string? get_gtk_menubar_object_path ();
+#if !HAS_MUTTER51
 		public unowned string? get_gtk_theme_variant ();
+#endif
 		public unowned string? get_gtk_unique_bus_name ();
 		public unowned string? get_gtk_window_object_path ();
 		public bool get_icon_geometry (out Mtk.Rectangle rect);
 		public uint64 get_id ();
 		public Meta.StackLayer get_layer ();
+#if HAS_MUTTER51
+		public bool get_max_size (out int width, out int height);
+#endif
 #if HAS_MUTTER49
 		public Meta.MaximizeFlags get_maximize_flags ();
 #else
 		public Meta.MaximizeFlags get_maximized ();
+#endif
+#if HAS_MUTTER51
+		public bool get_min_size (out int width, out int height);
 #endif
 		public int get_monitor ();
 		public unowned string? get_mutter_hints ();
@@ -1246,11 +1295,19 @@ namespace Meta {
 		[CCode (has_construct_function = false)]
 		protected WindowActor ();
 		public void freeze ();
+#if !HAS_MUTTER51
 		public Cairo.Surface? get_image (Mtk.Rectangle? clip);
+#endif
 		public unowned Meta.Window? get_meta_window ();
 		public unowned Meta.ShapedTexture? get_texture ();
 		public bool is_destroyed ();
+#if HAS_MUTTER51
+		public Cogl.Bitmap? paint_to_bitmap (Mtk.Rectangle? clip, Cogl.PixelFormat format);
+#endif
 		public Clutter.Content? paint_to_content (Mtk.Rectangle? clip) throws GLib.Error;
+#if HAS_MUTTER51
+		public Clutter.Content? paint_to_content_full (Mtk.Rectangle? clip, Cogl.PixelFormat format, Clutter.ColorState? color_state) throws GLib.Error;
+#endif
 		public void sync_visibility ();
 		public void thaw ();
 		public Meta.Window meta_window { get; construct; }
@@ -1907,6 +1964,9 @@ namespace Meta {
 	[CCode (cheader_filename = "meta/window.h", cprefix = "META_MAXIMIZE_", type_id = "meta_maximize_flags_get_type ()")]
 	[Flags]
 	public enum MaximizeFlags {
+#if HAS_MUTTER51
+		NONE,
+#endif
 		HORIZONTAL,
 		VERTICAL,
 		BOTH
diff --git a/vapi/meson.build b/vapi/meson.build
index 1e9695b4a..8266eaa33 100644
--- a/vapi/meson.build
+++ b/vapi/meson.build
@@ -401,3 +401,75 @@ if mutter50_dep.found()
 		output: 'libmutter-18.vapi'
 	)
 endif
+if mutter51_dep.found()
+	mtk_target = custom_target('mutter-mtk-51',
+		command: [
+			vapigen,
+			mutter_typelib_dir / 'Mtk-51.gir',
+			'--library=mutter-mtk-51',
+			'--pkg=gobject-2.0',
+			'--pkg=gio-2.0',
+			'--pkg=cairo',
+			'--pkg=graphene-gobject-1.0',
+			vapigen_args
+		],
+		output: 'mutter-mtk-51.vapi'
+	)
+	cogl_target = custom_target('mutter-cogl-51',
+		command: [
+			vapigen,
+			mutter_typelib_dir / 'Cogl-51.gir',
+			'--library=mutter-cogl-51',
+			'--pkg=mutter-mtk-51',
+			'--pkg=gobject-2.0',
+			'--pkg=gio-2.0',
+			'--pkg=cairo',
+			'--pkg=graphene-gobject-1.0',
+			'--pkg=x11',
+			vapigen_args,
+			files('Cogl-51-custom.vala')
+		],
+		depends: mtk_target,
+		output: 'mutter-cogl-51.vapi'
+	)
+
+	clutter_target = custom_target('mutter-clutter-51',
+		command: [
+			vapigen,
+			mutter_typelib_dir / 'Clutter-51.gir',
+			'--library=mutter-clutter-51',
+			'--pkg=graphene-gobject-1.0',
+			'--pkg=mutter-mtk-51',
+			'--pkg=mutter-cogl-51',
+			'--pkg=atk',
+			'--pkg=gio-2.0',
+			'--pkg=json-glib-1.0',
+			'--pkg=pango',
+			'--pkg=x11',
+			vapigen_args,
+		],
+		depends: [ cogl_target, mtk_target ],
+		output: 'mutter-clutter-51.vapi'
+	)
+
+	libmutter_target = custom_target('libmutter-51',
+		command: [
+			vapigen,
+			mutter_typelib_dir / 'Meta-51.gir',
+			'--library=libmutter-51',
+			'--pkg=graphene-gobject-1.0',
+			'--pkg=mutter-cogl-51',
+			'--pkg=mutter-clutter-51',
+			'--pkg=atk',
+			'--pkg=gio-2.0',
+			'--pkg=json-glib-1.0',
+			'--pkg=pango',
+			'--pkg=x11',
+			'--pkg=xfixes-4.0',
+			vapigen_args,
+			files('Meta-51-custom.vala')
+		],
+		depends: [ cogl_target, clutter_target ],
+		output: 'libmutter-51.vapi'
+	)
+endif
diff --git a/vapi/mutter-clutter-51.deps b/vapi/mutter-clutter-51.deps
new file mode 100644
index 000000000..9549905db
--- /dev/null
+++ b/vapi/mutter-clutter-51.deps
@@ -0,0 +1,6 @@
+atk
+cairo
+pango
+json-glib-1.0
+mutter-cogl-51
+graphene-gobject-1.0
diff --git a/vapi/mutter-clutter-51.vapi b/vapi/mutter-clutter-51.vapi
new file mode 120000
index 000000000..f71bfbc08
--- /dev/null
+++ b/vapi/mutter-clutter-51.vapi
@@ -0,0 +1 @@
+mutter-clutter-18.vapi
\ No newline at end of file
diff --git a/vapi/mutter-clutter.vapi b/vapi/mutter-clutter.vapi
index ef263d976..dfed083ba 100644
--- a/vapi/mutter-clutter.vapi
+++ b/vapi/mutter-clutter.vapi
@@ -5211,6 +5211,10 @@ namespace Clutter {
 		public Cogl.Color get_background_color ();
 #else
 		public Clutter.Color get_background_color ();
+#endif
+#if HAS_MUTTER51
+		[CCode (cname = "clutter_actor_class_get_binding_pool")]
+		public class unowned Clutter.BindingPool get_binding_pool ();
 #endif
 		public unowned Clutter.Actor? get_child_at_index (int index_);
 		public Graphene.Matrix get_child_transform ();
@@ -5664,6 +5668,9 @@ namespace Clutter {
 		[CCode (has_construct_function = false)]
 		protected Backend ();
 		public unowned Cogl.Context get_cogl_context ();
+#if HAS_MUTTER51
+		public Clutter.Cursor get_cursor (Clutter.CursorType cursor_type);
+#endif
 		public unowned Clutter.Seat get_default_seat ();
 #if !HAS_MUTTER48
 		public unowned Cairo.FontOptions get_font_options ();
@@ -5847,6 +5854,119 @@ namespace Clutter {
 		public ColorNode (Clutter.Color? color);
 #endif
 	}
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_color_op_get_type ()")]
+	public abstract class ColorOp : GLib.Object {
+		[CCode (has_construct_function = false)]
+		protected ColorOp ();
+		public void @3d_lut_get_data (uint32 out_size, float out_data);
+		public bool can_lower_to_3d_lut ();
+		public bool can_lower_to_curve_1d (size_t n_samples);
+		public bool can_lower_to_matrix_4x4 ();
+		public void curve_1d_get_data (size_t out_size, float out_r, float out_g, float out_b, float out_a);
+		public virtual void do_transform (float data, size_t n_samples);
+		public virtual float do_transform_one (float input);
+		public float gamma_power_get_power ();
+		public virtual bool get_clamps_input ();
+		public virtual bool get_clamps_output ();
+		public virtual bool get_transforms_alpha ();
+		public Clutter.ColorOp? lower_to_3d_lut (uint32 size);
+		public Clutter.ColorOp? lower_to_curve_1d (size_t n_samples);
+		public Clutter.ColorOp? lower_to_matrix_4x4 ();
+		public unowned Graphene.Matrix? matrix_4x4_get_matrix ();
+		public float multiply_get_value ();
+		public bool srgb_piecewise_eotf_get_unit_range_only ();
+		public bool srgb_piecewise_inv_eotf_get_unit_range_only ();
+		public virtual string to_string ();
+		public Clutter.ColorOp? try_combine (Clutter.ColorOp b);
+		public Clutter.YcbcrCoefficients ycbcr_matrix_get_coeffs ();
+		public unowned Graphene.Matrix? ycbcr_matrix_get_matrix ();
+	}
+	[CCode (cheader_filename = "clutter/clutter.h", lower_case_csuffix = "color_op_3d_lut", type_id = "clutter_color_op_3d_lut_get_type ()")]
+	public class ColorOp3DLut : Clutter.ColorOp {
+		[CCode (has_construct_function = false, type = "ClutterColorOp*")]
+		public ColorOp3DLut (uint32 size, float data);
+	}
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_color_op_clamp_unit_get_type ()")]
+	public class ColorOpClampUnit : Clutter.ColorOp {
+		[CCode (has_construct_function = false, type = "ClutterColorOp*")]
+		public ColorOpClampUnit ();
+	}
+	[CCode (cheader_filename = "clutter/clutter.h", lower_case_csuffix = "color_op_curve_1d", type_id = "clutter_color_op_curve_1d_get_type ()")]
+	public class ColorOpCurve1D : Clutter.ColorOp {
+		[CCode (has_construct_function = false, type = "ClutterColorOp*")]
+		public ColorOpCurve1D (size_t size, float r, float g, float b, float a);
+		[CCode (has_construct_function = false, type = "ClutterColorOp*")]
+		public ColorOpCurve1D.rgb (size_t size, float v);
+	}
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_color_op_gamma_power_get_type ()")]
+	public class ColorOpGammaPower : Clutter.ColorOp {
+		[CCode (has_construct_function = false, type = "ClutterColorOp*")]
+		public ColorOpGammaPower (float power);
+	}
+	[CCode (cheader_filename = "clutter/clutter.h", lower_case_csuffix = "color_op_matrix_4x4", type_id = "clutter_color_op_matrix_4x4_get_type ()")]
+	public class ColorOpMatrix4x4 : Clutter.ColorOp {
+		[CCode (has_construct_function = false, type = "ClutterColorOp*")]
+		public ColorOpMatrix4x4 (Graphene.Matrix matrix);
+	}
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_color_op_multiply_get_type ()")]
+	public class ColorOpMultiply : Clutter.ColorOp {
+		[CCode (has_construct_function = false, type = "ClutterColorOp*")]
+		public ColorOpMultiply (float value);
+	}
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_color_op_pq_eotf_get_type ()")]
+	public class ColorOpPqEotf : Clutter.ColorOp {
+		[CCode (has_construct_function = false, type = "ClutterColorOp*")]
+		public ColorOpPqEotf ();
+	}
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_color_op_pq_inv_eotf_get_type ()")]
+	public class ColorOpPqInvEotf : Clutter.ColorOp {
+		[CCode (has_construct_function = false, type = "ClutterColorOp*")]
+		public ColorOpPqInvEotf ();
+	}
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_color_op_premultiply_get_type ()")]
+	public class ColorOpPremultiply : Clutter.ColorOp {
+		[CCode (has_construct_function = false, type = "ClutterColorOp*")]
+		public ColorOpPremultiply ();
+	}
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_color_op_srgb_piecewise_eotf_get_type ()")]
+	public class ColorOpSrgbPiecewiseEotf : Clutter.ColorOp {
+		[CCode (has_construct_function = false, type = "ClutterColorOp*")]
+		public ColorOpSrgbPiecewiseEotf (bool unit_range_only);
+	}
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_color_op_srgb_piecewise_inv_eotf_get_type ()")]
+	public class ColorOpSrgbPiecewiseInvEotf : Clutter.ColorOp {
+		[CCode (has_construct_function = false, type = "ClutterColorOp*")]
+		public ColorOpSrgbPiecewiseInvEotf (bool unit_range_only);
+	}
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_color_op_unpremultiply_get_type ()")]
+	public class ColorOpUnpremultiply : Clutter.ColorOp {
+		[CCode (has_construct_function = false, type = "ClutterColorOp*")]
+		public ColorOpUnpremultiply ();
+	}
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_color_op_ycbcr_matrix_get_type ()")]
+	public class ColorOpYcbcrMatrix : Clutter.ColorOp {
+		[CCode (has_construct_function = false, type = "ClutterColorOp*")]
+		public ColorOpYcbcrMatrix (Clutter.YcbcrCoefficients coeffs, Graphene.Matrix matrix);
+	}
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_color_pipeline_get_type ()")]
+	public class ColorPipeline : GLib.Object {
+		[CCode (has_construct_function = false)]
+		protected ColorPipeline ();
+		public void add_op (Clutter.ColorOp op);
+		public void append (Clutter.ColorPipeline other);
+		public void combine ();
+		public void do_transform (float data, size_t n_samples);
+		public unowned GLib.List<Clutter.ColorOp> get_ops ();
+		public bool is_empty ();
+		public void shader_add_transform (Cogl.Pipeline cogl_pipeline);
+		public static bool shader_needs_color_state (Clutter.ColorState source_color_state, Clutter.ColorState target_color_state, Clutter.ColorStateTransformFlags flags);
+		public static void shader_set_color_state (Cogl.Pipeline cogl_pipeline, Clutter.ColorState source_color_state, Clutter.ColorState target_color_state, Clutter.ColorStateTransformFlags flags);
+		public void simplify ();
+		public void take_op (Clutter.ColorOp op);
+		public string to_string ();
+	}
+#endif
 	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_color_state_get_type ()")]
 	public sealed class ColorState : GLib.Object {
 		[CCode (has_construct_function = false)]
@@ -5856,12 +5976,15 @@ namespace Clutter {
 #else
 		public ColorState (Clutter.Context context, Clutter.Colorspace colorspace, Clutter.TransferFunction transfer_function);
 #endif
+#if !HAS_MUTTER51
 #if HAS_MUTTER49
 		public void add_pipeline_transform (Clutter.ColorState target_color_state, Cogl.Pipeline pipeline, Clutter.ColorStateTransformFlags flags);
 #else
 		public void add_pipeline_transform (Clutter.ColorState target_color_state, Cogl.Pipeline pipeline);
 #endif
+#endif
 #if HAS_MUTTER48
+#if !HAS_MUTTER51
 #if HAS_MUTTER49
 		[NoWrapper]
 		public virtual void append_transform_snippet (Clutter.ColorState target_color_state, GLib.StringBuilder snippet_globals, GLib.StringBuilder snippet_source, string snippet_color_var);
@@ -5872,6 +5995,7 @@ namespace Clutter {
 		public virtual void do_transform_to_XYZ ([CCode (array_length = false)] float[] data, int n_samples);
 #else
 		public virtual void do_transform (Clutter.ColorState target_color_state, [CCode (array_length = false)] float[] input, [CCode (array_length = false)] float[] output, int n_samples);
+#endif
 #endif
 		public virtual bool equals (Clutter.ColorState other_color_state);
 		public virtual Clutter.ColorState get_blending (bool force);
@@ -5890,6 +6014,7 @@ namespace Clutter {
 #if HAS_MUTTER47
 		public uint get_id ();
 #if HAS_MUTTER48
+#if !HAS_MUTTER51
 		[NoWrapper]
 #if HAS_MUTTER50
 		public virtual void init_color_transform_key (Clutter.ColorState target_color_state, Clutter.ColorStateTransformFlags flags, Clutter.ColorTransformKey key);
@@ -5897,6 +6022,7 @@ namespace Clutter {
 #else
 		public virtual void init_color_transform_key (Clutter.ColorState target_color_state, Clutter.ColorTransformKey key);
 #endif
+#endif
 #if HAS_MUTTER49
 		public void params_do_tone_mapping (Clutter.ColorState other_color_state, [CCode (array_length = false)] float[] data, int n_samples);
 #endif
@@ -5937,6 +6063,10 @@ namespace Clutter {
 		[CCode (has_construct_function = false, type = "ClutterColorState*")]
 		public ColorStateIcc (Clutter.Context context, uint8 icc_bytes, uint32 icc_length) throws GLib.Error;
 		public unowned Mtk.AnonymousFile get_file ();
+#if HAS_MUTTER51
+		public void* get_profile ();
+		public bool is_linearized ();
+#endif
 	}
 #endif
 #if HAS_MUTTER48
@@ -5960,10 +6090,18 @@ namespace Clutter {
 		public unowned Clutter.EOTF? get_eotf ();
 		public unowned Clutter.Luminance? get_luminance ();
 	}
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_color_transform_get_type ()")]
+	public sealed class ColorTransform : GLib.Object {
+		[CCode (has_construct_function = false)]
+		protected ColorTransform ();
+	}
+#else
 	[CCode (cheader_filename = "clutter/clutter.h", has_type_id = false)]
 	[Compact]
 	public class ColorTransformKey {
 	}
+#endif
 #endif
 	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_colorize_effect_get_type ()")]
 	public class ColorizeEffect : Clutter.OffscreenEffect {
@@ -5994,7 +6132,11 @@ namespace Clutter {
 		protected Context ();
 		public unowned Clutter.Backend get_backend ();
 #if HAS_MUTTER47
+#if HAS_MUTTER51
+		public unowned Clutter.ColorState get_default_color_state ();
+#else
 		public unowned Clutter.ColorManager get_color_manager ();
+#endif
 		public unowned Clutter.Settings get_settings ();
 #endif
 		public Clutter.TextDirection get_text_direction ();
@@ -6008,20 +6150,38 @@ namespace Clutter {
 	public abstract class Cursor : GLib.Object {
 		[CCode (has_construct_function = false)]
 		protected Cursor ();
+#if HAS_MUTTER51
+		public void emit_texture_changed ();
+#endif
 		[NoWrapper]
 		public virtual uint get_current_frame_time ();
+#if HAS_MUTTER51
+		public Clutter.CursorType get_cursor_type ();
+#endif
 		[NoWrapper]
 		public virtual void invalidate ();
 		[NoWrapper]
 		public virtual bool is_animated ();
 		[NoWrapper]
+#if HAS_MUTTER51
+		public virtual void prepare_at (float best_scale, float x, float y);
+#else
 		public virtual void prepare_at (float best_scale, int x, int y);
+#endif
 		[NoWrapper]
 		public virtual bool realize_texture ();
+#if HAS_MUTTER51
+		public void reset_viewport_dst_size ();
+		public void set_viewport_dst_size (int dst_width, int dst_height);
+#endif
 		[NoWrapper]
 		public virtual void tick_frame ();
 		[NoAccessorMethod]
 		public Clutter.ColorState color_state { construct; }
+#if HAS_MUTTER51
+		[NoAccessorMethod]
+		public Clutter.CursorType cursor_type { construct; }
+#endif
 		public signal void texture_changed ();
 	}
 #endif
@@ -6478,6 +6638,10 @@ namespace Clutter {
 	public abstract class InputFocus : GLib.Object {
 		[CCode (has_construct_function = false)]
 		protected InputFocus ();
+#if HAS_MUTTER51
+		[NoWrapper]
+		public virtual void action (Clutter.InputAction action);
+#endif
 		[NoWrapper]
 		public virtual void commit_text (string text);
 		[NoWrapper]
@@ -6498,7 +6662,11 @@ namespace Clutter {
 		public void set_cursor_location (Graphene.Rect rect);
 		public void set_input_panel_state (Clutter.InputPanelState state);
 		[NoWrapper]
+	#if HAS_MUTTER51
+		public virtual void set_preedit_text (string preedit, uint cursor, uint anchor, Clutter.PreeditAttribute style_hints, uint n_style_hints);
+	#else
 		public virtual void set_preedit_text (string preedit, uint cursor, uint anchor);
+	#endif
 		public void set_surrounding (string text, uint cursor, uint anchor);
 	}
 	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_input_method_get_type ()")]
@@ -6538,6 +6706,9 @@ namespace Clutter {
 		public Clutter.InputContentHintFlags content_hints { get; set; }
 		[NoAccessorMethod]
 		public Clutter.InputContentPurpose content_purpose { get; set; }
+#if HAS_MUTTER51
+		public Clutter.InputActionFlags handled_actions { set; }
+#endif
 		[HasEmitter]
 		public signal void commit (string text);
 		public signal void cursor_location_changed (Graphene.Rect object);
@@ -6576,6 +6747,23 @@ namespace Clutter {
 		public GLib.Value initial { owned get; set; }
 		public GLib.Type value_type { get; construct; }
 	}
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_key_controller_get_type ()")]
+	public sealed class KeyController : Clutter.Action {
+		[CCode (has_construct_function = false, type = "ClutterAction*")]
+		public KeyController (Clutter.InputFocus? im_focus);
+		public bool get_key (out uint32 key_symbol, out uint32 key_code, out unichar unicode);
+		public bool get_state (out Clutter.ModifierType pressed, out Clutter.ModifierType latched, out Clutter.ModifierType locked);
+		public bool get_trigger_keybindings ();
+		public void set_trigger_keybindings (bool trigger_keybindings);
+		[NoAccessorMethod]
+		public Clutter.InputFocus im_focus { construct; }
+		public bool trigger_keybindings { get; set; }
+		public signal bool key_press ();
+		public signal bool key_release ();
+		public signal void modifier_change ();
+	}
+#endif
 	[CCode (cheader_filename = "clutter/clutter.h", has_type_id = false)]
 	[Compact]
 	public class KeyEvent : Clutter.Event {
@@ -6662,6 +6850,16 @@ namespace Clutter {
 		[CCode (has_construct_function = false, type = "ClutterAction*")]
 		public LongPressGesture ();
 	}
+#endif
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_motion_controller_get_type ()")]
+	public class MotionController : Clutter.Action {
+		[CCode (has_construct_function = false, type = "ClutterAction*")]
+		public MotionController ();
+		public signal void enter (Clutter.Sprite sprite, float x, float y);
+		public signal void leave (Clutter.Sprite sprite);
+		public signal void motion (Clutter.Sprite sprite, float x, float y);
+	}
 #endif
 	[CCode (cheader_filename = "clutter/clutter.h", has_type_id = false)]
 	[Compact]
@@ -6830,6 +7028,7 @@ namespace Clutter {
 		public unowned Clutter.PickContext @ref ();
 		public void unref ();
 	}
+#if !HAS_MUTTER51
 #if HAS_MUTTER47
 	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_pipeline_cache_get_type ()")]
 	public class PipelineCache : GLib.Object {
@@ -6838,6 +7037,7 @@ namespace Clutter {
 		public void set_pipeline (Clutter.PipelineGroup group, int slot, Clutter.ColorState source_color_state, Clutter.ColorState target_color_state, Cogl.Pipeline pipeline);
 		public void unset_all_pipelines (Clutter.PipelineGroup group);
 	}
+#endif
 #endif
 	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_pipeline_node_get_type ()")]
 	public class PipelineNode : Clutter.PaintNode {
@@ -6924,6 +7124,16 @@ namespace Clutter {
 		public void set_scroll_mode (Clutter.ScrollMode mode);
 		public Clutter.ScrollMode scroll_mode { get; set; }
 	}
+#endif
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_scroll_controller_get_type ()")]
+	public class ScrollController : Clutter.Action {
+		[CCode (has_construct_function = false, type = "ClutterAction*")]
+		public ScrollController (Clutter.ScrollControllerFlags flags);
+		[NoAccessorMethod]
+		public Clutter.ScrollControllerFlags flags { get; construct; }
+		public signal void scroll (Clutter.Sprite sprite, Clutter.ScrollSource source, double dx, double dy);
+	}
 #endif
 	[CCode (cheader_filename = "clutter/clutter.h", has_type_id = false)]
 	[Compact]
@@ -7046,6 +7256,13 @@ namespace Clutter {
 	}
 	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_shader_effect_get_type ()")]
 	public class ShaderEffect : Clutter.OffscreenEffect {
+#if HAS_MUTTER51
+		[CCode (has_construct_function = false)]
+		protected ShaderEffect ();
+		[NoWrapper]
+		public virtual Cogl.Snippet get_static_snippet ();
+		public void set_uniform_float (string name, int n_components, [CCode (array_length_cname = "total_count", array_length_pos = 2.5)] float[] value);
+#else
 		[CCode (has_construct_function = false, type = "ClutterEffect*")]
 #if HAS_MUTTER48
 		public ShaderEffect (Cogl.ShaderType shader_type);
@@ -7057,12 +7274,18 @@ namespace Clutter {
 		[NoWrapper]
 		public virtual string get_static_shader_source ();
 		public bool set_shader_source (string source);
+#endif
 		public void set_uniform_value (string name, GLib.Value value);
+#if HAS_MUTTER51
+		[CCode (has_construct_function = false, type = "ClutterEffect*")]
+		public ShaderEffect.with_snippet (Cogl.Snippet snippet);
+#else
 		[NoAccessorMethod]
 #if HAS_MUTTER48
 		public Cogl.ShaderType shader_type { construct; }
 #else
 		public Clutter.ShaderType shader_type { construct; }
+#endif
 #endif
 	}
 	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_shader_float_get_type ()")]
@@ -7162,6 +7385,9 @@ namespace Clutter {
 		public Clutter.Content paint_to_content (Mtk.Rectangle rect, float scale, Clutter.PaintFlag paint_flags) throws GLib.Error;
 		public void paint_to_framebuffer (Cogl.Framebuffer framebuffer, Mtk.Rectangle rect, float scale, Clutter.PaintFlag paint_flags);
 #endif
+#if HAS_MUTTER51
+		public void paint_to_framebuffer_clipped (Cogl.Framebuffer framebuffer, Mtk.Rectangle rect, float scale, Clutter.ColorState color_state, Mtk.Region redraw_clip, Clutter.PaintFlag paint_flags);
+#endif
 #if !HAS_MUTTER49
 		public bool pointing_input_foreach (Clutter.StageInputForeachFunc func);
 #endif
@@ -7175,7 +7401,9 @@ namespace Clutter {
 		public void set_active (bool is_active);
 #endif
 		public void set_key_focus (Clutter.Actor? actor);
+#if !HAS_MUTTER51
 		public void set_minimum_size (uint width, uint height);
+#endif
 #if !HAS_MUTTER48
 		public void set_title (string title);
 		public void update_device (Clutter.InputDevice device, Clutter.EventSequence sequence, Clutter.InputDevice source_device, Graphene.Point point, uint32 time, Clutter.Actor new_actor, Mtk.Region region, bool emit_crossing);
@@ -7332,9 +7560,15 @@ namespace Clutter {
 		public bool get_cursor_visible ();
 		public bool get_editable ();
 		public Pango.EllipsizeMode get_ellipsize ();
+#if HAS_MUTTER51
+		public Cogl.Color get_error_color ();
+#endif
 		public Pango.FontDescription get_font_description ();
 		public unowned string get_font_name ();
 		public Clutter.InputContentHintFlags get_input_hints ();
+#if HAS_MUTTER51
+		public unowned Clutter.Actor get_input_interceptor ();
+#endif
 		public Clutter.InputContentPurpose get_input_purpose ();
 		public bool get_justify ();
 		public unowned Pango.Layout get_layout ();
@@ -7379,9 +7613,15 @@ namespace Clutter {
 		public void set_cursor_visible (bool cursor_visible);
 		public void set_editable (bool editable);
 		public void set_ellipsize (Pango.EllipsizeMode mode);
+#if HAS_MUTTER51
+		public void set_error_color (Cogl.Color? color);
+#endif
 		public void set_font_description (Pango.FontDescription font_desc);
 		public void set_font_name (string? font_name);
 		public void set_input_hints (Clutter.InputContentHintFlags hints);
+#if HAS_MUTTER51
+		public void set_input_interceptor (Clutter.Actor? input_interceptor);
+#endif
 		public void set_input_purpose (Clutter.InputContentPurpose purpose);
 		public void set_justify (bool justify);
 		public void set_line_alignment (Pango.Alignment alignment);
@@ -7430,9 +7670,17 @@ namespace Clutter {
 		public bool cursor_visible { get; set; }
 		public bool editable { get; set; }
 		public Pango.EllipsizeMode ellipsize { get; set; }
+#if HAS_MUTTER51
+		public Cogl.Color error_color { get; set; }
+		[NoAccessorMethod]
+		public bool error_color_set { get; }
+#endif
 		public Pango.FontDescription font_description { owned get; set; }
 		public string font_name { get; set; }
 		public Clutter.InputContentHintFlags input_hints { get; set; }
+#if HAS_MUTTER51
+		public Clutter.Actor input_interceptor { get; set; }
+#endif
 		public Clutter.InputContentPurpose input_purpose { get; set; }
 		public bool justify { get; set; }
 		public Pango.Alignment line_alignment { get; set; }
@@ -7752,13 +8000,26 @@ namespace Clutter {
 		public Clutter.ColorimetryType type;
 		public Clutter.Colorspace colorspace;
 		public Clutter.Primaries primaries;
+#if HAS_MUTTER51
+		public bool equal (Clutter.Colorimetry other_colorimetry);
+		public void from_XYZ (Graphene.Matrix from_XYZ);
+		public unowned Clutter.Primaries? get_primaries ();
+		public void to_XYZ (Graphene.Matrix to_XYZ);
+		public string to_string ();
+#endif
 	}
 	[CCode (cheader_filename = "clutter/clutter.h", has_type_id = false)]
 	public struct EOTF {
 		public Clutter.EOTFType type;
 		public Clutter.TransferFunction tf_name;
 		public float gamma_exp;
+#if HAS_MUTTER51
+		public bool equal (Clutter.EOTF other_eotf);
+#endif
 		public unowned Clutter.Luminance? get_default_luminance ();
+#if HAS_MUTTER51
+		public unowned string to_string ();
+#endif
 	}
 	[CCode (cheader_filename = "clutter/clutter.h", has_type_id = false)]
 	public struct Luminance {
@@ -7768,6 +8029,10 @@ namespace Clutter {
 		public float @ref;
 #if HAS_MUTTER50
 		public float mastering_max;
+#endif
+#if HAS_MUTTER51
+		public bool equal (Clutter.Luminance other_lum);
+		public static unowned Clutter.Luminance? get_default_sdr ();
 #endif
 	}
 #endif
@@ -7835,6 +8100,9 @@ namespace Clutter {
 		public float get_height ();
 		public Graphene.Point3D get_origin ();
 		public float get_width ();
+#if HAS_MUTTER51
+		public void scale (float x_scale, float y_scale, float z_scale);
+#endif
 		public void set_depth (float depth);
 		public bool set_from_allocation (Clutter.Actor actor);
 		public void set_height (float height);
@@ -7889,6 +8157,9 @@ namespace Clutter {
 		public float w_x;
 		public float w_y;
 		public void ensure_normalized_range ();
+#if HAS_MUTTER51
+		public bool equal (Clutter.Primaries other_primaries);
+#endif
 	}
 #endif
 #if HAS_MUTTER47
@@ -8217,6 +8488,9 @@ namespace Clutter {
 	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_PHASE_", type_id = "clutter_event_phase_get_type ()")]
 	public enum EventPhase {
 		CAPTURE,
+#if HAS_MUTTER51
+		TARGET,
+#endif
 		BUBBLE
 	}
 	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_", type_id = "clutter_event_type_get_type ()")]
@@ -8389,6 +8663,10 @@ namespace Clutter {
 		HIDDEN_TEXT,
 		SENSITIVE_DATA,
 		LATIN,
+#if HAS_MUTTER51
+		INHIBIT_OSK,
+		NO_EMOJI,
+#endif
 		MULTILINE
 	}
 	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_INPUT_CONTENT_PURPOSE_", type_id = "clutter_input_content_purpose_get_type ()")]
@@ -8646,6 +8924,17 @@ namespace Clutter {
 		NEAREST,
 		TRILINEAR
 	}
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_SCROLL_CONTROLLER_FLAG_", type_id = "clutter_scroll_controller_flags_get_type ()")]
+	[Flags]
+	public enum ScrollControllerFlags {
+		NONE,
+		SCROLL_VERTICAL,
+		SCROLL_HORIZONTAL,
+		DISCRETE,
+		PHYSICAL_DIRECTION
+	}
+#endif
 	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_SCROLL_", type_id = "clutter_scroll_direction_get_type ()")]
 	public enum ScrollDirection {
 		UP,
@@ -8770,6 +9059,18 @@ namespace Clutter {
 		POINTER,
 		TOUCHSCREEN
 	}
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_YCBCR_COEFFICIENTS_", type_id = "clutter_ycbcr_coefficients_get_type ()")]
+	public enum YcbcrCoefficients {
+		IDENTITY_LIMITED,
+		BT601_FULL,
+		BT601_LIMITED,
+		BT709_FULL,
+		BT709_LIMITED,
+		BT2020_FULL,
+		BT2020_LIMITED
+	}
+#endif
 	[CCode (cheader_filename = "clutter/clutter.h", instance_pos = 1.9)]
 	public delegate Clutter.Actor ActorCreateChildFunc (GLib.Object item);
 	[CCode (cheader_filename = "clutter/clutter.h", instance_pos = 2.9)]
@@ -8863,4 +9164,8 @@ namespace Clutter {
 #endif
 	[CCode (cheader_filename = "clutter/clutter.h")]
 	public static uint unicode_to_keysym (uint32 wc);
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "clutter/clutter.h")]
+	public static void xyY_to_XYZ (float x, float y, float Y, Graphene.Vec3 XYZ);
+#endif
 }
diff --git a/vapi/mutter-cogl-14.vapi b/vapi/mutter-cogl-14.vapi
index b0306668e..a64a0f779 100644
--- a/vapi/mutter-cogl-14.vapi
+++ b/vapi/mutter-cogl-14.vapi
@@ -90,6 +90,12 @@ namespace Cogl {
 		public int get_height ();
 		public int get_rowstride ();
 		public int get_width ();
+#if HAS_MUTTER51
+		public unowned uint8? map (Cogl.BufferAccess access, Cogl.BufferMapHint hints) throws GLib.Error;
+		public void unmap ();
+		[CCode (has_construct_function = false)]
+		public Bitmap.with_malloc_buffer (Cogl.Context context, uint width, uint height, Cogl.PixelFormat format) throws GLib.Error;
+#endif
 		[CCode (has_construct_function = false)]
 		public Bitmap.with_size (Cogl.Context context, uint width, uint height, Cogl.PixelFormat format);
 	}
@@ -113,9 +119,17 @@ namespace Cogl {
 		public Cogl.BufferUpdateHint update_hint { set construct; }
 	}
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_context_get_type ()")]
+#if HAS_MUTTER51
+	public abstract class Context : GLib.Object {
+#else
 	public sealed class Context : GLib.Object {
+#endif
 		[CCode (has_construct_function = false)]
+#if HAS_MUTTER51
+		protected Context ();
+#else
 		public Context (Cogl.Display? display) throws GLib.Error;
+#endif
 #if HAS_MUTTER47
 		public void flush ();
 #else
@@ -159,14 +173,35 @@ namespace Cogl {
 		public bool is_hardware_accelerated ();
 #endif
 		public void set_named_pipeline (Cogl.PipelineKey key, Cogl.Pipeline? pipeline);
+#if HAS_MUTTER51
+		public void set_winsys_feature (Cogl.WinsysFeature feature, bool value);
+#endif
 #if !HAS_MUTTER50
 		public int64 timestamp_query_get_time_ns (Cogl.TimestampQuery query);
+#endif
+#if HAS_MUTTER51
+		public Cogl.Display display { get; construct; }
 #endif
 	}
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_context_egl_get_type ()")]
+	public class ContextEGL : Cogl.Context {
+		[CCode (has_construct_function = false, type = "CoglContext*")]
+		public ContextEGL (Cogl.Display display) throws GLib.Error;
+	}
+#endif
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_display_get_type ()")]
+#if HAS_MUTTER51
+	public abstract class Display : GLib.Object {
+#else
 	public sealed class Display : GLib.Object {
+#endif
 		[CCode (has_construct_function = false)]
-#if HAS_MUTTER47
+#if HAS_MUTTER51
+		protected Display ();
+		[NoWrapper]
+		public virtual void destroy ();
+#elif HAS_MUTTER47
 		public Display (Cogl.Renderer renderer);
 #else
 		public Display (Cogl.Renderer renderer, Cogl.OnscreenTemplate onscreen_template);
@@ -176,7 +211,21 @@ namespace Cogl {
 		public void set_onscreen_template (Cogl.OnscreenTemplate onscreen_template);
 #endif
 		public bool setup () throws GLib.Error;
+#if HAS_MUTTER51
+		public Cogl.Renderer renderer { get; construct; }
+#endif
 	}
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_display_egl_get_type ()")]
+	public class DisplayEGL : Cogl.Display {
+		[CCode (has_construct_function = false)]
+		public DisplayEGL (Cogl.Renderer renderer);
+		[NoWrapper]
+		public virtual void cleanup_context ();
+		[NoWrapper]
+		public virtual bool context_created () throws GLib.Error;
+	}
+#endif
 	[CCode (cheader_filename = "cogl/cogl.h", free_function = "cogl_dma_buf_handle_free", has_type_id = false)]
 	[Compact]
 	public class DmaBufHandle {
@@ -281,6 +330,9 @@ namespace Cogl {
 		[CCode (cheader_filename = "cogl/cogl.h", cname = "cogl_blit_framebuffer")]
 #endif
 		public bool blit (Cogl.Framebuffer dst, int src_x, int src_y, int dst_x, int dst_y, int width, int height) throws GLib.Error;
+#if HAS_MUTTER51
+		public bool blit_region (Cogl.Framebuffer dst, Mtk.Region src_region, int dst_x_offset, int dst_y_offset) throws GLib.Error;
+#endif
 #if !HAS_MUTTER47
 		public void cancel_fence_callback (Cogl.FenceClosure closure);
 #endif
@@ -477,7 +529,12 @@ namespace Cogl {
 		public void show ();
 #endif
 		public void swap_buffers (Cogl.FrameInfo frame_info, void* user_data);
-#if HAS_MUTTER48
+#if HAS_MUTTER51
+		[NoWrapper]
+		public virtual bool swap_buffers_with_damage (Mtk.Region region, Cogl.FrameInfo info);
+		[NoWrapper]
+		public virtual bool swap_region (Mtk.Region region, Cogl.FrameInfo info);
+#elif HAS_MUTTER48
 		public virtual void swap_buffers_with_damage (Mtk.Region region, Cogl.FrameInfo info);
 		public virtual void swap_region (Mtk.Region region, Cogl.FrameInfo info);
 #else
@@ -522,6 +579,9 @@ namespace Cogl {
 	public sealed class Pipeline : GLib.Object {
 		[CCode (has_construct_function = false)]
 		public Pipeline (Cogl.Context context);
+#if HAS_MUTTER51
+		public void add_capability (GLib.Quark domain, uint capability);
+#endif
 		public void add_layer_snippet (int layer, Cogl.Snippet snippet);
 		public void add_snippet (Cogl.Snippet snippet);
 		public Cogl.Pipeline copy ();
@@ -544,7 +604,9 @@ namespace Cogl {
 		public bool get_per_vertex_point_size ();
 		public float get_point_size ();
 		public int get_uniform_location (string uniform_name);
+#if !HAS_MUTTER51
 		public unowned Cogl.Program get_user_program ();
+#endif
 #if HAS_MUTTER48
 		public bool has_capability (GLib.Quark domain, uint capability);
 #endif
@@ -583,7 +645,9 @@ namespace Cogl {
 		public void set_uniform_int (int uniform_location, int n_components, int count, int value);
 		public void set_uniform_matrix (int uniform_location, int dimensions, int count, bool transpose, float value);
 #endif
+#if !HAS_MUTTER51
 		public void set_user_program (Cogl.Program program);
+#endif
 	}
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_pixel_buffer_get_type ()")]
 	public sealed class PixelBuffer : Cogl.Buffer {
@@ -640,6 +704,7 @@ namespace Cogl {
 		[CCode (has_construct_function = false)]
 		public Primitive.with_attributes (Cogl.VerticesMode mode, int n_vertices, [CCode (array_length_cname = "n_attributes", array_length_pos = 3.1)] Cogl.Attribute[] attributes);
 	}
+#if !HAS_MUTTER51
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_program_get_type ()")]
 	public sealed class Program : GLib.Object {
 		[CCode (has_construct_function = false)]
@@ -667,10 +732,17 @@ namespace Cogl {
 		[Version (deprecated = true, deprecated_since = "1.16")]
 		public void set_uniform_matrix (int uniform_location, int dimensions, bool transpose, [CCode (array_length_cname = "count", array_length_pos = 2.5)] float[] value);
 	}
+#endif
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_renderer_get_type ()")]
+#if HAS_MUTTER51
+	public abstract class Renderer : GLib.Object {
+		[CCode (has_construct_function = false)]
+		protected Renderer ();
+#else
 	public sealed class Renderer : GLib.Object {
 		[CCode (has_construct_function = false)]
 		public Renderer ();
+#endif
 #if !HAS_MUTTER47
 		public void add_constraint (Cogl.RendererConstraint constraint);
 #endif
@@ -696,6 +768,10 @@ namespace Cogl {
 #if HAS_MUTTER47
 		public void* get_proc_address (string name);
 #endif
+#if HAS_MUTTER51
+		public virtual int get_sync_fd ();
+#endif
+#if !HAS_MUTTER51
 #if HAS_MUTTER50
 		public unowned Cogl.Winsys get_winsys ();
 #elif HAS_MUTTER49
@@ -705,6 +781,7 @@ namespace Cogl {
 		public void* get_winsys_data ();
 #else
 		public Cogl.WinsysID get_winsys_id ();
+#endif
 #endif
 		public bool is_dma_buf_supported ();
 #if HAS_MUTTER48
@@ -719,6 +796,15 @@ namespace Cogl {
 #if !HAS_MUTTER47
 		public void set_winsys_id (Cogl.WinsysID winsys_id);
 #endif
+#if HAS_MUTTER51
+		[NoWrapper]
+		public virtual bool load_driver (Cogl.DriverId driver_id) throws GLib.Error;
+		public void set_driver_id (Cogl.DriverId driver);
+		[NoWrapper]
+		public virtual void update_sync ();
+		[NoAccessorMethod]
+		public Cogl.Driver driver { owned get; construct; }
+#else
 #if HAS_MUTTER50
 		public void set_custom_winsys (Cogl.Winsys winsys);
 #endif
@@ -729,8 +815,17 @@ namespace Cogl {
 #else
 		public void set_winsys (void* winsys);
 #endif
+#endif
 #endif
 	}
+#if HAS_MUTTER51
+	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_renderer_egl_get_type ()")]
+	public class RendererEGL : Cogl.Renderer {
+		[CCode (has_construct_function = false)]
+		public RendererEGL ();
+		public bool has_feature (Cogl.EGLWinsysFeature feature);
+	}
+#endif
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_scanout_get_type ()")]
 	public sealed class Scanout : GLib.Object {
 		[CCode (has_construct_function = false)]
@@ -754,6 +849,7 @@ namespace Cogl {
 		public void set_src_rect (Graphene.Rect rect);
 		public signal void scanout_failed (Cogl.Onscreen object);
 	}
+#if !HAS_MUTTER51
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_shader_get_type ()")]
 	public sealed class Shader : GLib.Object {
 		[CCode (has_construct_function = false)]
@@ -771,10 +867,14 @@ namespace Cogl {
 		[Version (deprecated = true, deprecated_since = "1.16")]
 		public void source (string source);
 	}
+#endif
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_snippet_get_type ()")]
 	public sealed class Snippet : GLib.Object {
 		[CCode (has_construct_function = false)]
 		public Snippet (Cogl.SnippetHook hook, string? declarations, string? post);
+#if HAS_MUTTER51
+		public bool get_capability (out GLib.Quark domain, out uint capability);
+#endif
 		public unowned string get_declarations ();
 		public Cogl.SnippetHook get_hook ();
 		public unowned string get_post ();
@@ -881,6 +981,7 @@ namespace Cogl {
 		[CCode (has_construct_function = false, type = "CoglTexture*")]
 		public Texture2DSliced.with_size (Cogl.Context ctx, int width, int height, int max_waste);
 	}
+#if !HAS_MUTTER51
 #if HAS_MUTTER50
 	[CCode (cheader_filename = "cogl/cogl.h", type_id = "cogl_winsys_get_type ()")]
 	public abstract class Winsys : GLib.Object {
@@ -921,6 +1022,7 @@ namespace Cogl {
 	[Compact]
 	public class TimestampQuery {
 	}
+#endif
 #endif
 	[CCode (cheader_filename = "cogl/cogl.h", type_cname = "CoglScanoutBufferInterface", type_id = "cogl_scanout_buffer_get_type ()")]
 	public interface ScanoutBuffer : GLib.Object {
@@ -989,7 +1091,7 @@ namespace Cogl {
 		public void set_test_function (Cogl.DepthTestFunction function);
 		public void set_write_enabled (bool enable);
 	}
-#if HAS_MUTTER50
+#if HAS_MUTTER50 && !HAS_MUTTER51
 	[CCode (cheader_filename = "cogl/cogl.h", has_type_id = false)]
 	public struct DisplayEGL {
 		public void* platform;
@@ -1023,7 +1125,7 @@ namespace Cogl {
 		public Cogl.Color color;
 	}
 #endif
-#if HAS_MUTTER50
+#if HAS_MUTTER50 && !HAS_MUTTER51
 	[CCode (cheader_filename = "cogl/cogl.h", has_type_id = false)]
 	public struct RendererEGL {
 		public Cogl.EGLWinsysFeature private_features;
@@ -1240,8 +1342,10 @@ namespace Cogl {
 		OGL_FEATURE_ID_MAP_BUFFER_FOR_READ,
 		[CCode (cname = "COGL_FEATURE_ID_MAP_BUFFER_FOR_WRITE")]
 		OGL_FEATURE_ID_MAP_BUFFER_FOR_WRITE,
+#if !HAS_MUTTER51
 		[CCode (cname = "COGL_FEATURE_ID_FENCE")]
 		OGL_FEATURE_ID_FENCE,
+#endif
 		[CCode (cname = "COGL_FEATURE_ID_TEXTURE_RG")]
 		OGL_FEATURE_ID_TEXTURE_RG,
 		[CCode (cname = "COGL_FEATURE_ID_TEXTURE_RGBA1010102")]
@@ -1254,8 +1358,10 @@ namespace Cogl {
 		[CCode (cname = "COGL_FEATURE_ID_BUFFER_AGE")]
 		OGL_FEATURE_ID_BUFFER_AGE,
 #endif
+#if !HAS_MUTTER51
 		[CCode (cname = "COGL_FEATURE_ID_TEXTURE_EGL_IMAGE_EXTERNAL")]
 		OGL_FEATURE_ID_TEXTURE_EGL_IMAGE_EXTERNAL,
+#endif
 		[CCode (cname = "COGL_FEATURE_ID_BLIT_FRAMEBUFFER")]
 		OGL_FEATURE_ID_BLIT_FRAMEBUFFER,
 #if HAS_MUTTER47 && !HAS_MUTTER50
@@ -1373,7 +1479,11 @@ namespace Cogl {
 	}
 	[CCode (cheader_filename = "cogl/cogl.h", cprefix = "COGL_PIXEL_FORMAT_", type_id = "cogl_pixel_format_get_type ()")]
 	public enum PixelFormat {
+#if HAS_MUTTER51
+        ARGB32_NATIVE,
+#else
 		CAIRO_ARGB32_COMPAT,
+#endif
 		ANY,
 		A_8,
 		RGB_565,
@@ -1467,6 +1577,7 @@ namespace Cogl {
 		EGL
 	}
 #endif
+#if !HAS_MUTTER51
 #if HAS_MUTTER48
 	[CCode (cheader_filename = "cogl/cogl.h", cprefix = "COGL_SHADER_TYPE_", type_id = "cogl_shader_type_get_type ()")]
 #else
@@ -1476,6 +1587,7 @@ namespace Cogl {
 		VERTEX,
 		FRAGMENT
 	}
+#endif
 	[CCode (cheader_filename = "cogl/cogl.h", cprefix = "COGL_SNIPPET_HOOK_", has_type_id = false)]
 	public enum SnippetHook {
 		VERTEX,
@@ -1563,7 +1675,11 @@ namespace Cogl {
 		SWAP_REGION_SYNCHRONIZED,
 #endif
 		BUFFER_AGE,
+#if HAS_MUTTER51
+		SYNC_FD,
+#else
 		SYNC_AND_COMPLETE_EVENT,
+#endif
 		N_FEATURES
 	}
 #if !HAS_MUTTER50
@@ -1626,8 +1742,15 @@ namespace Cogl {
 	[CCode (cheader_filename = "cogl/cogl.h", cprefix = "COGL_SYSTEM_ERROR_", has_type_id = false)]
 #endif
 	public errordomain SystemError {
+#if HAS_MUTTER51
+		[CCode (cname = "COGL_SYSTEM_ERROR_UNSUPPORTED")]
+		COGL_SYSTEM_ERROR_UNSUPPORTED,
+		[CCode (cname = "COGL_SYSTEM_ERROR_NO_MEMORY")]
+		COGL_SYSTEM_ERROR_NO_MEMORY
+#else
 		UNSUPPORTED,
 		NO_MEMORY
+#endif
 	}
 	[CCode (cheader_filename = "cogl/cogl.h", cprefix = "COGL_TEXTURE_ERROR_", has_type_id = false)]
 	public errordomain TextureError {
diff --git a/vapi/mutter-cogl-51.deps b/vapi/mutter-cogl-51.deps
new file mode 100644
index 000000000..35fbfd0ab
--- /dev/null
+++ b/vapi/mutter-cogl-51.deps
@@ -0,0 +1,5 @@
+pango
+glib-2.0
+gio-2.0
+mutter-mtk-51
+x11
diff --git a/vapi/mutter-cogl-51.vapi b/vapi/mutter-cogl-51.vapi
new file mode 120000
index 000000000..e85846834
--- /dev/null
+++ b/vapi/mutter-cogl-51.vapi
@@ -0,0 +1 @@
+mutter-cogl-18.vapi
\ No newline at end of file
diff --git a/vapi/mutter-mtk-14.vapi b/vapi/mutter-mtk-14.vapi
index 29815d2b2..2e2c59591 100644
--- a/vapi/mutter-mtk-14.vapi
+++ b/vapi/mutter-mtk-14.vapi
@@ -172,6 +172,8 @@ namespace Mtk {
 #if HAS_MUTTER50
 	[CCode (cheader_filename = "mtk/mtk.h")]
 	public static int64 find_nearest_interval_boundary (int64 boundary_us, int64 base_us, int64 interval_us);
+	[CCode (cheader_filename = "mtk/mtk.h")]
+	public static bool is_fd_readable (int fd);
 #endif
 	[CCode (cheader_filename = "mtk/mtk.h")]
 	[Version (replacement = "Rectangle.from_graphene_rect")]
diff --git a/vapi/mutter-mtk-51.vapi b/vapi/mutter-mtk-51.vapi
new file mode 120000
index 000000000..914519b0e
--- /dev/null
+++ b/vapi/mutter-mtk-51.vapi
@@ -0,0 +1 @@
+mutter-mtk-18.vapi
\ No newline at end of file

From f50a4bc1372496c3c0575cb63640c8b26a47f7bd Mon Sep 17 00:00:00 2001
From: Denis Garaev <garaevdi@yandex.ru>
Date: Sun, 12 Jul 2026 15:21:09 +0300
Subject: [PATCH 3/5] Shaders: use `Cogl.Snippet`

---
 lib/Effects/BoxBlurManager.vala                 | 16 ++++++++++++++++
 lib/Effects/RoundedCornersEffect.vala           | 17 +++++++++++++++++
 .../ColorblindnessCorrectionEffect.vala         | 16 ++++++++++++++++
 src/ColorFilters/MonochromeEffect.vala          | 15 +++++++++++++++
 4 files changed, 64 insertions(+)

diff --git a/lib/Effects/BoxBlurManager.vala b/lib/Effects/BoxBlurManager.vala
index e42d8aaca..41123cf97 100644
--- a/lib/Effects/BoxBlurManager.vala
+++ b/lib/Effects/BoxBlurManager.vala
@@ -35,21 +35,30 @@ public class Gala.BoxBlurManager : Object {
 
         private const float[] HORIZONTAL_PASS_DATA = { 1.0f, 0.0f };
         private const float[] VERTICAL_PASS_DATA = { 0.0f, 1.0f };
+#if HAS_MUTTER51
+        private string shader;
+#endif
 
         public int radius { set { set_uniform_value ("RADIUS", value); } }
 
         public BoxBlurEffect (PassDirection direction) {
             Object (
+#if !HAS_MUTTER51
 #if HAS_MUTTER48
                 shader_type: Cogl.ShaderType.FRAGMENT
 #else
                 shader_type: Clutter.ShaderType.FRAGMENT_SHADER
+#endif
 #endif
             );
 
             try {
                 var bytes = GLib.resources_lookup_data ("/io/elementary/desktop/gala/shaders/box-blur.frag", NONE);
+#if HAS_MUTTER51
+                shader = (string) bytes.get_data ();
+#else
                 set_shader_source ((string) bytes.get_data ());
+#endif
             } catch (Error e) {
                 critical ("Unable to load box-blur.frag: %s", e.message);
             }
@@ -63,6 +72,13 @@ public class Gala.BoxBlurManager : Object {
             set_uniform_value ("DIRECTION", direction_value);
         }
 
+#if HAS_MUTTER51
+        public override Cogl.Snippet get_static_snippet () {
+            // TODO: split declarations from shader code and put it here
+            return new Cogl.Snippet (Cogl.SnippetHook.FRAGMENT, null, shader);
+        }
+#endif
+
         public override void set_actor (Clutter.Actor? new_actor) {
             if (actor != null) {
                 actor.notify["width"].disconnect (update_pixel_step);
diff --git a/lib/Effects/RoundedCornersEffect.vala b/lib/Effects/RoundedCornersEffect.vala
index 641339bf6..49a338c39 100644
--- a/lib/Effects/RoundedCornersEffect.vala
+++ b/lib/Effects/RoundedCornersEffect.vala
@@ -33,12 +33,18 @@ public class Gala.RoundedCornersEffect : Clutter.ShaderEffect {
         }
     }
 
+#if HAS_MUTTER51
+    private string shader;
+#endif
+
     public RoundedCornersEffect (int clip_radius, float monitor_scale) {
         Object (
+#if !HAS_MUTTER51
 #if HAS_MUTTER48
             shader_type: Cogl.ShaderType.FRAGMENT,
 #else
             shader_type: Clutter.ShaderType.FRAGMENT_SHADER,
+#endif
 #endif
             clip_radius: clip_radius,
             monitor_scale: monitor_scale
@@ -48,12 +54,23 @@ public class Gala.RoundedCornersEffect : Clutter.ShaderEffect {
     construct {
         try {
             var bytes = GLib.resources_lookup_data ("/io/elementary/desktop/gala/shaders/rounded-corners.frag", NONE);
+#if HAS_MUTTER51
+            shader = (string) bytes.get_data ();
+#else
             set_shader_source ((string) bytes.get_data ());
+#endif
         } catch (Error e) {
             critical ("Unable to load rounded-corners.frag: %s", e.message);
         }
     }
 
+#if HAS_MUTTER51
+    public override Cogl.Snippet get_static_snippet () {
+        // TODO: split declarations from shader code and put it here
+        return new Cogl.Snippet (Cogl.SnippetHook.FRAGMENT, null, shader);
+    }
+#endif
+
     public override void set_actor (Clutter.Actor? new_actor) {
         if (actor != null) {
             actor.notify["width"].disconnect (update_actor_size);
diff --git a/src/ColorFilters/ColorblindnessCorrectionEffect.vala b/src/ColorFilters/ColorblindnessCorrectionEffect.vala
index 4cb5416f4..df02247a4 100644
--- a/src/ColorFilters/ColorblindnessCorrectionEffect.vala
+++ b/src/ColorFilters/ColorblindnessCorrectionEffect.vala
@@ -29,6 +29,9 @@ public class Gala.ColorblindnessCorrectionEffect : Clutter.ShaderEffect {
             queue_repaint ();
         }
     }
+#if HAS_MUTTER51
+    private string shader;
+#endif
 
     /*
      * Used for fading in and out the effect, since you can't add transitions to effects.
@@ -37,10 +40,12 @@ public class Gala.ColorblindnessCorrectionEffect : Clutter.ShaderEffect {
 
     public ColorblindnessCorrectionEffect (int mode, double strength) {
         Object (
+#if !HAS_MUTTER51
 #if HAS_MUTTER48
             shader_type: Cogl.ShaderType.FRAGMENT,
 #else
             shader_type: Clutter.ShaderType.FRAGMENT_SHADER,
+#endif
 #endif
             mode: mode,
             strength: strength
@@ -48,11 +53,22 @@ public class Gala.ColorblindnessCorrectionEffect : Clutter.ShaderEffect {
 
         try {
             var bytes = GLib.resources_lookup_data ("/io/elementary/desktop/gala/shaders/colorblindness-correction.frag", GLib.ResourceLookupFlags.NONE);
+#if HAS_MUTTER51
+            shader = (string) bytes.get_data ();
+#else
             set_shader_source ((string) bytes.get_data ());
+#endif
         } catch (Error e) {
             critical ("Unable to load colorblindness-correction.frag: %s", e.message);
         }
 
         pause_for_screenshot = false;
     }
+
+#if HAS_MUTTER51
+    public override Cogl.Snippet get_static_snippet () {
+        // TODO: split declarations from shader code and put it here
+        return new Cogl.Snippet (Cogl.SnippetHook.FRAGMENT, null, shader);
+    }
+#endif
 }
diff --git a/src/ColorFilters/MonochromeEffect.vala b/src/ColorFilters/MonochromeEffect.vala
index c5b8034d7..a51370284 100644
--- a/src/ColorFilters/MonochromeEffect.vala
+++ b/src/ColorFilters/MonochromeEffect.vala
@@ -21,6 +21,9 @@ public class Gala.MonochromeEffect : Clutter.ShaderEffect {
             queue_repaint ();
         }
     }
+#if HAS_MUTTER51
+    private string shader;
+#endif
 
     /*
      * Used for fading in and out the effect, since you can't add transitions to effects.
@@ -29,21 +32,33 @@ public class Gala.MonochromeEffect : Clutter.ShaderEffect {
 
     public MonochromeEffect (double strength) {
         Object (
+#if !HAS_MUTTER51
 #if HAS_MUTTER48
             shader_type: Cogl.ShaderType.FRAGMENT,
 #else
             shader_type: Clutter.ShaderType.FRAGMENT_SHADER,
+#endif
 #endif
             strength: strength
         );
 
         try {
             var bytes = GLib.resources_lookup_data ("/io/elementary/desktop/gala/shaders/monochrome.frag", GLib.ResourceLookupFlags.NONE);
+#if HAS_MUTTER51
+            shader = (string) bytes.get_data ();
+#else
             set_shader_source ((string) bytes.get_data ());
+#endif
         } catch (Error e) {
             critical ("Unable to load monochrome.frag: %s", e.message);
         }
 
         pause_for_screenshot = false;
     }
+#if HAS_MUTTER51
+    public override Cogl.Snippet get_static_snippet () {
+        // TODO: split declarations from shader code and put it here
+        return new Cogl.Snippet (Cogl.SnippetHook.FRAGMENT, null, shader);
+    }
+#endif
 }

From dc58f78ea4df2e3cd1335282791165c806c5ca82 Mon Sep 17 00:00:00 2001
From: Denis Garaev <garaevdi@yandex.ru>
Date: Sun, 30 Aug 2026 23:10:46 +0300
Subject: [PATCH 4/5] Canvas: update PixelFormat

---
 lib/Drawing/Canvas.vala | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/lib/Drawing/Canvas.vala b/lib/Drawing/Canvas.vala
index 3e52f083c..da3ae975f 100644
--- a/lib/Drawing/Canvas.vala
+++ b/lib/Drawing/Canvas.vala
@@ -34,7 +34,11 @@ public class Gala.Drawing.Canvas : GLib.Object, Clutter.Content {
         int real_width = (int) Math.ceilf (width * scale_factor);
         int real_height = (int) Math.ceilf (height * scale_factor);
         if (bitmap == null) {
+#if HAS_MUTTER51
+            bitmap = new Cogl.Bitmap.with_size (cogl_context, real_width, real_height, Cogl.PixelFormat.ARGB32_NATIVE);
+#else
             bitmap = new Cogl.Bitmap.with_size (cogl_context, real_width, real_height, Cogl.PixelFormat.CAIRO_ARGB32_COMPAT);
+#endif
         }
 
         unowned Cogl.Buffer? buffer = bitmap.get_buffer ();

From e9e62f758c41123df4095c9eba6011efeac5caf4 Mon Sep 17 00:00:00 2001
From: Denis Garaev <garaevdi@yandex.ru>
Date: Sun, 30 Aug 2026 23:13:16 +0300
Subject: [PATCH 5/5] ScreenshotManager: use new Meta.WindowActor API

Inspired by
https://gitlab.gnome.org/GNOME/gnome-shell/-/commit/e7a5d3836514677c61568158630cd0762af56015
---
 src/Misc/ScreenshotManager.vala | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/src/Misc/ScreenshotManager.vala b/src/Misc/ScreenshotManager.vala
index 27f982c9a..ebaa6281b 100644
--- a/src/Misc/ScreenshotManager.vala
+++ b/src/Misc/ScreenshotManager.vala
@@ -365,7 +365,16 @@ public class Gala.ScreenshotManager : Object {
         var resource_scale = window_actor.get_resource_scale ();
 
         Mtk.Rectangle main_clip = { main_rect.x - (int) window_actor.x, main_rect.y - (int) window_actor.y, main_rect.width, main_rect.height };
+#if HAS_MUTTER51
+        var bitmap = window_actor.paint_to_bitmap (null, Cogl.PixelFormat.ARGB32_NATIVE);
+        var bitmap_width = bitmap.get_width ();
+        var bitmap_height = bitmap.get_height ();
+        var bitmap_stride = bitmap.get_rowstride ();
+        var data = bitmap.map (Cogl.BufferAccess.READ, Cogl.BufferMapHint.DISCARD);
+        var window_image = new Cairo.ImageSurface.for_data ((uchar[]) data, Cairo.Format.ARGB32, bitmap_width, bitmap_height, bitmap_stride);
+#else
         var window_image = (Cairo.ImageSurface) window_actor.get_image (main_clip);
+#endif
 
         var image = new Cairo.ImageSurface (
             ARGB32,
@@ -391,7 +400,16 @@ public class Gala.ScreenshotManager : Object {
                 transient_rect.width,
                 transient_rect.height
             };
+#if HAS_MUTTER51
+            var transient_bitmap = transient_actor.paint_to_bitmap (transient_clip, Cogl.PixelFormat.ARGB32_NATIVE);
+            var transient_bitmap_width = bitmap.get_width ();
+            var transient_bitmap_height = bitmap.get_height ();
+            var transient_bitmap_stride = bitmap.get_rowstride ();
+            var transient_data = transient_bitmap.map (Cogl.BufferAccess.READ, Cogl.BufferMapHint.DISCARD);
+            var transient_image = new Cairo.ImageSurface.for_data ((uchar[]) transient_data, Cairo.Format.ARGB32, transient_bitmap_width, transient_bitmap_height, transient_bitmap_stride);
+#else
             var transient_image = (Cairo.ImageSurface) transient_actor.get_image (transient_clip);
+#endif
 
             cairo_context.set_source_surface (
                 transient_image,
