diff --git a/lib/CloseButton.vala b/lib/CloseButton.vala
index 2f9e5d8f..afc7a7b2 100644
--- a/lib/CloseButton.vala
+++ b/lib/CloseButton.vala
@@ -11,7 +11,11 @@ public class Gala.CloseButton : Clutter.Actor {
     public float monitor_scale { get; construct set; }
 
     private Icon icon;
+#if HAS_MUTTER49
+    private Clutter.ClickGesture click_action;
+#else
     private Clutter.ClickAction click_action;
+#endif
 
     public CloseButton (float monitor_scale) {
         Object (monitor_scale: monitor_scale);
@@ -28,9 +32,17 @@ public class Gala.CloseButton : Clutter.Actor {
         };
         add_child (icon);
 
+#if HAS_MUTTER49
+        click_action = new Clutter.ClickGesture ();
+#else
         click_action = new Clutter.ClickAction ();
+#endif
         add_action (click_action);
+#if HAS_MUTTER49
+        click_action.recognize.connect (on_clicked);
+#else
         click_action.clicked.connect (on_clicked);
+#endif
         click_action.notify["pressed"].connect (on_pressed_changed);
     }
 
diff --git a/lib/DragDropAction.vala b/lib/DragDropAction.vala
index e72b62d9..8b04692a 100644
--- a/lib/DragDropAction.vala
+++ b/lib/DragDropAction.vala
@@ -246,7 +246,7 @@ namespace Gala {
                     clicked = false;
                     dragging = true;
 
-                    grab_actor (handle, event.get_device ());
+                    grab_actor (handle, event.get_source_device ());
 
                     var source_list = sources.@get (drag_id);
                     if (source_list != null) {
@@ -316,7 +316,7 @@ namespace Gala {
         }
 
         private bool on_event (Clutter.Event event) requires (dragging) {
-            var device = event.get_device ();
+            var device = event.get_source_device ();
 
             if (grabbed_device != null &&
                 device != grabbed_device &&
diff --git a/lib/Gestures/TouchpadBackend.vala b/lib/Gestures/TouchpadBackend.vala
index 34303b15..4546cb21 100644
--- a/lib/Gestures/TouchpadBackend.vala
+++ b/lib/Gestures/TouchpadBackend.vala
@@ -91,7 +91,7 @@ private class Gala.TouchpadBackend : Object, GestureBackend {
 
             gesture.type = event.get_type ();
             gesture.fingers = (int) event.get_touchpad_gesture_finger_count ();
-            gesture.performed_on_device_type = event.get_device ().get_device_type ();
+            gesture.performed_on_device_type = event.get_source_device ().get_device_type ();
 
             if (!on_gesture_detected (gesture, event.get_time ())) {
                 if (state == NONE) {
diff --git a/src/ShellClients/HideTracker.vala b/src/ShellClients/HideTracker.vala
index 8065921a..0ffd4cc0 100644
--- a/src/ShellClients/HideTracker.vala
+++ b/src/ShellClients/HideTracker.vala
@@ -18,7 +18,11 @@ public class Gala.HideTracker : Object {
 
     private static GLib.Settings behavior_settings;
 
+#if HAS_MUTTER49
+    private Clutter.PanGesture pan_action;
+#else
     private Clutter.PanAction pan_action;
+#endif
 
     private bool hovered = false;
 
@@ -60,13 +64,22 @@ public class Gala.HideTracker : Object {
         });
 
         display.window_created.connect (on_window_created);
-
+#if HAS_MUTTER49
+        pan_action = new Clutter.PanGesture () {
+            min_n_points = 1,
+            max_n_points = 1,
+            pan_axis = Clutter.PanAxis.X
+        };
+        pan_action.may_recognize.connect (check_valid_gesture);
+        pan_action.pan_update.connect (on_pan);
+#else
         pan_action = new Clutter.PanAction () {
             n_touch_points = 1,
             pan_axis = X_AXIS
         };
         pan_action.gesture_begin.connect (check_valid_gesture);
         pan_action.pan.connect (on_pan);
+#endif
 
 #if HAS_MUTTER48
         display.get_compositor ().get_stage ().add_action_full ("panel-swipe-gesture", CAPTURE, pan_action);
@@ -136,7 +149,11 @@ public class Gala.HideTracker : Object {
         }
 
         float y;
+#if HAS_MUTTER49
+        y = pan_action.get_point_begin_coords (0).y;
+#else
         pan_action.get_press_coords (0, null, out y);
+#endif
 
         var monitor_geom = display.get_monitor_geometry (panel.window.get_monitor ());
         if ((y - monitor_geom.y - monitor_geom.height).abs () < 50) { // Only start if the gesture starts near the bottom of the monitor
@@ -146,16 +163,30 @@ public class Gala.HideTracker : Object {
         return false;
     }
 
+#if HAS_MUTTER49
+    private void on_pan () {
+#else
     private bool on_pan () {
+#endif
         float delta_y;
+#if HAS_MUTTER49
+        delta_y = pan_action.get_delta ().get_y ();
+#else
         pan_action.get_motion_delta (0, null, out delta_y);
+#endif
 
         if (delta_y < 0) { // Only allow swipes upwards
+#if HAS_MUTTER49
+            panel.window.focus (pan_action.get_point_event (0).get_time ());
+#else
             panel.window.focus (pan_action.get_last_event (0).get_time ());
+#endif
             trigger_show ();
         }
 
+#if !HAS_MUTTER49
         return false;
+#endif
     }
 
     private void setup_barrier () {
diff --git a/src/Widgets/MultitaskingView/WindowClone.vala b/src/Widgets/MultitaskingView/WindowClone.vala
index e36fb83b..ab41e565 100644
--- a/src/Widgets/MultitaskingView/WindowClone.vala
+++ b/src/Widgets/MultitaskingView/WindowClone.vala
@@ -112,9 +112,14 @@ public class Gala.WindowClone : ActorTarget, RootTarget {
         window.size_changed.connect (() => request_reposition ());
 
         if (mode != MULTITASKING_VIEW) {
+#if HAS_MUTTER49
+            var click_action = new Clutter.ClickGesture ();
+            click_action.recognize.connect ((action) => {
+#else
             var click_action = new Clutter.ClickAction ();
             click_action.clicked.connect ((action, actor) => {
-                actor_clicked (action.get_button ());
+#endif
+                actor_clicked (((Clutter.PressGesture) action).get_button ());
             });
 
             add_action (click_action);
diff --git a/src/Widgets/MultitaskingView/WorkspaceClone.vala b/src/Widgets/MultitaskingView/WorkspaceClone.vala
index 2b13d65f..d57b17d1 100644
--- a/src/Widgets/MultitaskingView/WorkspaceClone.vala
+++ b/src/Widgets/MultitaskingView/WorkspaceClone.vala
@@ -132,8 +132,13 @@ public class Gala.WorkspaceClone : ActorTarget {
         unowned var display = workspace.get_display ();
         var monitor_geometry = display.get_monitor_geometry (display.get_primary_monitor ());
 
+#if HAS_MUTTER49
+        var background_click_action = new Clutter.ClickGesture ();
+        background_click_action.recognize.connect (() => activate (true));
+#else
         var background_click_action = new Clutter.ClickAction ();
         background_click_action.clicked.connect (() => activate (true));
+#endif
         background = new FramedBackground (display);
         background.add_action (background_click_action);
 
diff --git a/src/WindowManager.vala b/src/WindowManager.vala
index ee6e1959..ae208100 100644
--- a/src/WindowManager.vala
+++ b/src/WindowManager.vala
@@ -754,8 +754,10 @@ namespace Gala {
                     event.get_type () == TOUCHPAD_HOLD || event.get_type () == TOUCH_BEGIN) {
                     window.begin_grab_op (
                         op,
-                        event.get_device (),
+                        null,
+#if !HAS_MUTTER49
                         event.get_event_sequence (),
+#endif
                         event.get_time ()
 #if HAS_MUTTER46
                         , null
diff --git a/vapi/Clutter-17.metadata b/vapi/Clutter-17.metadata
index 4b7be62e..8c22eeeb 100644
--- a/vapi/Clutter-17.metadata
+++ b/vapi/Clutter-17.metadata
@@ -123,10 +123,6 @@ TouchpadHoldEvent struct=false base_type="Clutter.Event"
 TouchpadPinchEvent struct=false base_type="Clutter.Event"
 TouchpadSwipeEvent struct=false base_type="Clutter.Event"
 
-Focus skip
-KeyFocus skip
-Sprite skip
-
 // Keysyms used to be CLUTTER_X instead of CLUTTER_KEY_X
 *#constant skip
 CURRENT_TIME skip=false
diff --git a/vapi/libmutter.vapi b/vapi/libmutter.vapi
index 967dd161..98d41dc0 100644
--- a/vapi/libmutter.vapi
+++ b/vapi/libmutter.vapi
@@ -991,7 +991,10 @@ namespace Meta {
 		public void activate_with_workspace (uint32 current_time, Meta.Workspace workspace);
 		public bool allows_move ();
 		public bool allows_resize ();
-#if HAS_MUTTER46
+#if HAS_MUTTER49
+		// Here instead of an object should be a sprite, but adding it would also drag Clutter.Focus in
+		public bool begin_grab_op (Meta.GrabOp op, GLib.Object? _, uint32 timestamp, Graphene.Point? pos_hint);
+#elif HAS_MUTTER46
 		public bool begin_grab_op (Meta.GrabOp op, Clutter.InputDevice? device, Clutter.EventSequence? sequence, uint32 timestamp, Graphene.Point? pos_hint);
 #else
 		public bool begin_grab_op (Meta.GrabOp op, Clutter.InputDevice? device, Clutter.EventSequence? sequence, uint32 timestamp);
diff --git a/vapi/mutter-clutter.vapi b/vapi/mutter-clutter.vapi
index b8766955..d384bd04 100644
--- a/vapi/mutter-clutter.vapi
+++ b/vapi/mutter-clutter.vapi
@@ -5811,10 +5811,17 @@ namespace Clutter {
 		public Clutter.Container container { get; construct; }
 	}
 #endif
+#if HAS_MUTTER49
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_click_gesture_get_type ()")]
+	public sealed class ClickGesture : Clutter.PressGesture {
+		[CCode (has_construct_function = false, type = "ClutterAction*")]
+		public ClickGesture ();
+#else
 	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_click_action_get_type ()")]
 	public class ClickAction : Clutter.Action {
 		[CCode (has_construct_function = false, type = "ClutterAction*")]
 		public ClickAction ();
+#endif
 		public uint get_button ();
 		public void get_coords (out float press_x, out float press_y);
 		public Clutter.ModifierType get_state ();
@@ -6770,6 +6777,36 @@ namespace Clutter {
 #endif
 		public void unref ();
 	}
+#if HAS_MUTTER49
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_pan_gesture_get_type ()")]
+	public sealed class PanGesture : Clutter.Gesture {
+		[CCode (has_construct_function = false, type = "ClutterAction*")]
+		public PanGesture ();
+		public Graphene.Vec2 get_accumulated_delta ();
+		public Graphene.Vec2 get_accumulated_delta_abs ();
+		public Graphene.Point get_begin_centroid ();
+		public Graphene.Point get_begin_centroid_abs ();
+		public uint get_begin_threshold ();
+		public Graphene.Point get_centroid ();
+		public Graphene.Point get_centroid_abs ();
+		public Graphene.Vec2 get_delta ();
+		public Graphene.Vec2 get_delta_abs ();
+		public uint get_max_n_points ();
+		public uint get_min_n_points ();
+		public Clutter.PanAxis get_pan_axis ();
+		public Graphene.Vec2 get_velocity ();
+		public Graphene.Vec2 get_velocity_abs ();
+		public void set_begin_threshold (uint begin_threshold);
+		public void set_max_n_points (uint max_n_points);
+		public void set_min_n_points (uint min_n_points);
+		public void set_pan_axis (Clutter.PanAxis axis);
+		public uint begin_threshold { get; set; }
+		public uint max_n_points { get; set; }
+		public uint min_n_points { get; set; }
+		public Clutter.PanAxis pan_axis { get; set; }
+		public signal void pan_update ();
+	}
+#else
 	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_pan_action_get_type ()")]
 	public class PanAction : Clutter.GestureAction {
 		[CCode (has_construct_function = false, type = "ClutterAction*")]
@@ -6794,6 +6831,7 @@ namespace Clutter {
 		public signal bool pan (Clutter.Actor actor, bool is_interpolated);
 		public virtual signal void pan_stopped (Clutter.Actor actor);
 	}
+#endif
 #if !HAS_MUTTER46
 	[CCode (cheader_filename = "clutter/clutter.h", lower_case_csuffix = "param_units", type_id = "clutter_param_units_get_type ()")]
 	public class ParamSpecUnit : GLib.ParamSpec {
@@ -6871,6 +6909,36 @@ namespace Clutter {
 		[CCode (has_construct_function = false, type = "ClutterPaintNode*")]
 		public PipelineNode (Cogl.Pipeline? pipeline);
 	}
+#if HAS_MUTTER49
+	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_press_gesture_get_type ()")]
+	public abstract class PressGesture : Clutter.Gesture {
+		[CCode (has_construct_function = false)]
+		protected PressGesture ();
+		public uint get_button ();
+		public int get_cancel_threshold ();
+		public Graphene.Point get_coords ();
+		public Graphene.Point get_coords_abs ();
+		public int get_long_press_duration_ms ();
+		public uint get_n_presses ();
+		public bool get_pressed ();
+		public uint get_required_button ();
+		public Clutter.ModifierType get_state ();
+		[NoWrapper]
+		public virtual void long_press ();
+		[NoWrapper]
+		public virtual void press ();
+		[NoWrapper]
+		public virtual void release ();
+		public void set_cancel_threshold (int cancel_threshold);
+		public void set_long_press_duration_ms (int long_press_duration_ms);
+		public void set_required_button (uint required_button);
+		public int cancel_threshold { get; set; }
+		public int long_press_duration_ms { get; set; }
+		[NoAccessorMethod]
+		public bool pressed { get; set; }
+		public uint required_button { get; set; }
+	}
+#endif
 	[CCode (cheader_filename = "clutter/clutter.h", type_id = "clutter_property_transition_get_type ()")]
 #if HAS_MUTTER46
 	public class PropertyTransition : Clutter.Transition {
@@ -8610,12 +8678,20 @@ namespace Clutter {
 		FORCE_CURSORS,
 		CLEAR
 	}
+#if HAS_MUTTER49
+	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_PAN_AXIS_", type_id = "clutter_pan_axis_get_type ()")]
+	public enum PanAxis {
+		BOTH,
+		X,
+		Y
+#else
 	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_PAN_", type_id = "clutter_pan_axis_get_type ()")]
 	public enum PanAxis {
 		AXIS_NONE,
 		X_AXIS,
 		Y_AXIS,
 		AXIS_AUTO
+#endif
 	}
 #if !HAS_MUTTER46
 	[CCode (cheader_filename = "clutter/clutter.h", cprefix = "CLUTTER_PATH_", type_id = "clutter_path_node_type_get_type ()")]
