diff --git a/src/XdgDesktopPortalPantheon.vala b/src/XdgDesktopPortalPantheon.vala
index b83a1ac..084eb6a 100644
--- a/src/XdgDesktopPortalPantheon.vala
+++ b/src/XdgDesktopPortalPantheon.vala
@@ -66,8 +66,12 @@ private void on_name_acquired () {
     // appearing before binding the style scheme
     uint watch_id = 0;
     watch_id = Bus.watch_name (BusType.SESSION, "org.freedesktop.portal.Desktop", BusNameWatcherFlags.NONE, () => {
+        Gtk.init ();
         Granite.init ();
 
+        weak Gtk.IconTheme default_theme = Gtk.IconTheme.get_for_display (Gdk.Display.get_default ());
+        default_theme.add_resource_path ("/io/elementary/xdg-desktop-portal-pantheon/icons");
+
         var css_provider = new Gtk.CssProvider ();
         css_provider.load_from_resource ("/io/elementary/xdg-desktop-portal-pantheon/Application.css");
         Gtk.StyleContext.add_provider_for_display (Gdk.Display.get_default (), css_provider, Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION);
@@ -83,6 +87,7 @@ private void on_name_acquired () {
         if (watch_id != 0) {
             Bus.unwatch_name (watch_id);
         }
+        debug ("Granite set up complete");
     });
 
 }
@@ -96,11 +101,6 @@ int main (string[] args) {
     /* Avoid pointless and confusing recursion */
     GLib.Environment.unset_variable ("GTK_USE_PORTAL");
 
-    Gtk.init ();
-
-    weak Gtk.IconTheme default_theme = Gtk.IconTheme.get_for_display (Gdk.Display.get_default ());
-    default_theme.add_resource_path ("/io/elementary/xdg-desktop-portal-pantheon/icons");
-
     try {
         var opt_context = new OptionContext ("- portal backends");
         opt_context.set_summary ("A backend implementation for xdg-desktop-portal.");
