diff --git a/src/Services/DBusInterfaces/Screen.vala b/src/Services/DBusInterfaces/Screen.vala
index 102faee0..2a61fcac 100644
--- a/src/Services/DBusInterfaces/Screen.vala
+++ b/src/Services/DBusInterfaces/Screen.vala
@@ -18,8 +18,8 @@
  */
 
 namespace Power.Services.DBusInterfaces {
-  [DBus (name = "org.gnome.SettingsDaemon.Power.Screen")]
-  interface PowerSettings : GLib.Object {
+  [DBus (name = "io.elementary.gala.BrightnessManager")]
+  interface BrightnessManager : GLib.Object {
     public abstract int brightness { get; set; }
   }
 }
diff --git a/src/Services/DeviceManager.vala b/src/Services/DeviceManager.vala
index 7ce3eb8e..806fe4df 100644
--- a/src/Services/DeviceManager.vala
+++ b/src/Services/DeviceManager.vala
@@ -24,10 +24,13 @@ public class Power.Services.DeviceManager : Object {
     private const string POWER_SETTINGS_INTERFACE = "org.gnome.SettingsDaemon.Power";
     private const string POWER_SETTINGS_PATH = "/org/gnome/SettingsDaemon/Power";
 
+    private const string GALA_INTERFACE = "io.elementary.gala";
+    private const string GALA_PATH = "/io/elementary/gala/BrightnessManager";
+
     private static DeviceManager? instance = null;
 
     private DBusInterfaces.UPower? upower = null;
-    private DBusInterfaces.PowerSettings? iscreen = null;
+    private DBusInterfaces.BrightnessManager? iscreen = null;
 
     public Services.Backlight backlight { get; construct; }
     public Gee.HashMap<string, Device> devices { get; private set; }
@@ -92,8 +95,8 @@ public class Power.Services.DeviceManager : Object {
 
             iscreen = yield Bus.get_proxy (
                 BusType.SESSION,
-                POWER_SETTINGS_INTERFACE,
-                POWER_SETTINGS_PATH,
+                GALA_INTERFACE,
+                GALA_PATH,
                 DBusProxyFlags.GET_INVALIDATED_PROPERTIES
             );
             debug ("Connection to Power Settings bus established");
